{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc04b16e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "DATA_DIR = Path(\"/home/carlyn.1/dna-trait-analysis/data\")\n",
    "\n",
    "melpomene_reference_info_file = DATA_DIR / \"Hmel2.5.gff3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c42d80c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "/home/carlyn.1/dna-trait-analysis/tmp/mummer-4.0.1/nucmer -c 100 -p nucmer /home/carlyn.1/dna-trait-analysis/data/Heliconius_melpomene_melpomene_Hmel2.5.scaffolds.fa /home/carlyn.1/dna-trait-analysis/data/Heliconius_erato_demophoon_v1_-_scaffolds.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d1f51ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "/home/carlyn.1/dna-trait-analysis/tmp/mummer-4.0.1/show-coords -r -c -l nucmer.delta > nucmer.coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "372b5d3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "/home/carlyn.1/dna-trait-analysis/tmp/mummer-4.0.1/show-snps -C nucmer.delta > nucmer.snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2279dd30",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "/home/carlyn.1/dna-trait-analysis/tmp/mummer-4.0.1/show-tiling nucmer.delta > nucmer.tiling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b9355f57",
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = []\n",
    "with open(\"nucmer.coords\", \"r\") as f:\n",
    "    for line in f.readlines()[5:]:\n",
    "        # Seems to be inclusive (the start and end positions)\n",
    "        (\n",
    "            mel_start,\n",
    "            mel_end,\n",
    "            _,\n",
    "            era_start,\n",
    "            era_end,\n",
    "            _,\n",
    "            mel_len,\n",
    "            era_len,\n",
    "            _,\n",
    "            idy,\n",
    "            _,\n",
    "            len_r,\n",
    "            len_q,\n",
    "            _,\n",
    "            cov_r,\n",
    "            cov_q,\n",
    "            _,\n",
    "            mel_tag,\n",
    "            era_tag,\n",
    "        ) = line.split()\n",
    "\n",
    "        erato_chrom_info = era_tag.replace(\"Herato\", \"\")\n",
    "        if erato_chrom_info in [\"_mt\"]:\n",
    "            continue\n",
    "        erato_chrom_num = int(erato_chrom_info[:2])\n",
    "        erato_scaf_num = int(erato_chrom_info[2:])\n",
    "        melpomene_chrom_info = mel_tag.replace(\"Hmel2\", \"\")\n",
    "        melpomene_chrom_num = int(melpomene_chrom_info[:2])\n",
    "        melpomene_scaf_num = int(melpomene_chrom_info[2:-1])\n",
    "\n",
    "        real_mel_start = min(mel_start, mel_end)\n",
    "        real_mel_end = max(mel_start, mel_end)\n",
    "        real_era_start = min(era_start, era_end)\n",
    "        real_era_end = max(era_start, era_end)\n",
    "\n",
    "        rows.append(\n",
    "            [\n",
    "                real_mel_start,\n",
    "                real_mel_end,\n",
    "                real_era_start,\n",
    "                real_era_end,\n",
    "                mel_len,\n",
    "                era_len,\n",
    "                idy,\n",
    "                len_r,\n",
    "                len_q,\n",
    "                cov_r,\n",
    "                cov_q,\n",
    "                mel_tag,\n",
    "                era_tag,\n",
    "                melpomene_chrom_num,\n",
    "                melpomene_scaf_num,\n",
    "                erato_chrom_num,\n",
    "                erato_scaf_num,\n",
    "            ]\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "aa585854",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>melpomene_start</th>\n",
       "      <th>melpomene_end</th>\n",
       "      <th>erato_start</th>\n",
       "      <th>erato_end</th>\n",
       "      <th>melpomene_length</th>\n",
       "      <th>erato_length</th>\n",
       "      <th>idy</th>\n",
       "      <th>len_r</th>\n",
       "      <th>len_q</th>\n",
       "      <th>cov_r</th>\n",
       "      <th>cov_q</th>\n",
       "      <th>melpomene_tag</th>\n",
       "      <th>erato_tag</th>\n",
       "      <th>melpomene_chromosome</th>\n",
       "      <th>melpomene_scaffold</th>\n",
       "      <th>erato_chromosome</th>\n",
       "      <th>erato_scaffold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8004</th>\n",
       "      <td>10000142</td>\n",
       "      <td>9998677</td>\n",
       "      <td>13297916</td>\n",
       "      <td>13299351</td>\n",
       "      <td>1466</td>\n",
       "      <td>1436</td>\n",
       "      <td>83.47</td>\n",
       "      <td>17206585</td>\n",
       "      <td>22325789</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Hmel201001o</td>\n",
       "      <td>Herato0101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8005</th>\n",
       "      <td>10000267</td>\n",
       "      <td>10000572</td>\n",
       "      <td>13299872</td>\n",
       "      <td>13300176</td>\n",
       "      <td>306</td>\n",
       "      <td>305</td>\n",
       "      <td>88.75</td>\n",
       "      <td>17206585</td>\n",
       "      <td>22325789</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Hmel201001o</td>\n",
       "      <td>Herato0101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8006</th>\n",
       "      <td>10000953</td>\n",
       "      <td>10002162</td>\n",
       "      <td>13300580</td>\n",
       "      <td>13301816</td>\n",
       "      <td>1210</td>\n",
       "      <td>1237</td>\n",
       "      <td>85.26</td>\n",
       "      <td>17206585</td>\n",
       "      <td>22325789</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Hmel201001o</td>\n",
       "      <td>Herato0101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8007</th>\n",
       "      <td>10003128</td>\n",
       "      <td>10004074</td>\n",
       "      <td>13302673</td>\n",
       "      <td>13303589</td>\n",
       "      <td>947</td>\n",
       "      <td>917</td>\n",
       "      <td>84.12</td>\n",
       "      <td>17206585</td>\n",
       "      <td>22325789</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Hmel201001o</td>\n",
       "      <td>Herato0101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8008</th>\n",
       "      <td>10006179</td>\n",
       "      <td>10006473</td>\n",
       "      <td>13306123</td>\n",
       "      <td>13306431</td>\n",
       "      <td>295</td>\n",
       "      <td>309</td>\n",
       "      <td>89.07</td>\n",
       "      <td>17206585</td>\n",
       "      <td>22325789</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Hmel201001o</td>\n",
       "      <td>Herato0101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     melpomene_start melpomene_end erato_start erato_end melpomene_length  \\\n",
       "8004        10000142       9998677    13297916  13299351             1466   \n",
       "8005        10000267      10000572    13299872  13300176              306   \n",
       "8006        10000953      10002162    13300580  13301816             1210   \n",
       "8007        10003128      10004074    13302673  13303589              947   \n",
       "8008        10006179      10006473    13306123  13306431              295   \n",
       "\n",
       "     erato_length    idy     len_r     len_q cov_r cov_q melpomene_tag  \\\n",
       "8004         1436  83.47  17206585  22325789  0.01  0.01   Hmel201001o   \n",
       "8005          305  88.75  17206585  22325789  0.00  0.00   Hmel201001o   \n",
       "8006         1237  85.26  17206585  22325789  0.01  0.01   Hmel201001o   \n",
       "8007          917  84.12  17206585  22325789  0.01  0.00   Hmel201001o   \n",
       "8008          309  89.07  17206585  22325789  0.00  0.00   Hmel201001o   \n",
       "\n",
       "       erato_tag  melpomene_chromosome  melpomene_scaffold  erato_chromosome  \\\n",
       "8004  Herato0101                     1                   1                 1   \n",
       "8005  Herato0101                     1                   1                 1   \n",
       "8006  Herato0101                     1                   1                 1   \n",
       "8007  Herato0101                     1                   1                 1   \n",
       "8008  Herato0101                     1                   1                 1   \n",
       "\n",
       "      erato_scaffold  \n",
       "8004               1  \n",
       "8005               1  \n",
       "8006               1  \n",
       "8007               1  \n",
       "8008               1  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "alignment_coords_df = pd.DataFrame(\n",
    "    rows,\n",
    "    columns=[\n",
    "        \"melpomene_start\",\n",
    "        \"melpomene_end\",\n",
    "        \"erato_start\",\n",
    "        \"erato_end\",\n",
    "        \"melpomene_length\",\n",
    "        \"erato_length\",\n",
    "        \"idy\",\n",
    "        \"len_r\",\n",
    "        \"len_q\",\n",
    "        \"cov_r\",\n",
    "        \"cov_q\",\n",
    "        \"melpomene_tag\",\n",
    "        \"erato_tag\",\n",
    "        \"melpomene_chromosome\",\n",
    "        \"melpomene_scaffold\",\n",
    "        \"erato_chromosome\",\n",
    "        \"erato_scaffold\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "alignment_coords_df = alignment_coords_df[\n",
    "    alignment_coords_df.melpomene_chromosome == alignment_coords_df.erato_chromosome\n",
    "]  # Only keep alignments on the same chromosome\n",
    "alignment_coords_df = alignment_coords_df.sort_values(\n",
    "    by=[\"melpomene_chromosome\", \"melpomene_scaffold\", \"melpomene_start\"]\n",
    ")\n",
    "alignment_coords_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "56a4a307",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(158208, 17)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment_coords_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "9c532bfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "rows = []\n",
    "other_names = [\"chromosome\", \"scaffold\"]\n",
    "att_names = [\n",
    "    \"Dbxref\",\n",
    "    \"ID\",\n",
    "    \"Name\",\n",
    "    \"Note\",\n",
    "    \"Parent\",\n",
    "    \"deletions\",\n",
    "    \"description\",\n",
    "    \"frameshifts\",\n",
    "    \"stable_id\",\n",
    "    \"symbol\",\n",
    "    \"translation_stable_id\",\n",
    "]\n",
    "with open(melpomene_reference_info_file, \"r\") as f:\n",
    "    for line in f.readlines():\n",
    "        cols = line.split(\"\\t\")\n",
    "        if len(cols) == 1:\n",
    "            continue  # Just a seperator line\n",
    "\n",
    "        seqid, source, feat_type, start, end, score, strand, phase, attributes = cols\n",
    "        atts = attributes.strip().split(\";\")\n",
    "        att_cols = [\"\" for _ in range(len(att_names))]\n",
    "        for att in atts:\n",
    "            parts = att.split(\"=\")\n",
    "            key = parts[0]\n",
    "            if key not in att_names:\n",
    "                continue\n",
    "            value = parts[1]\n",
    "            att_idx = att_names.index(key)\n",
    "            att_cols[att_idx] = value\n",
    "\n",
    "            chrom_info = seqid.replace(\"Hmel2\", \"\")\n",
    "            chrom_num = int(chrom_info[:2])\n",
    "            scaf_num = int(chrom_info[2:-1])\n",
    "\n",
    "        other_cols = [chrom_num, scaf_num]\n",
    "        rows.append(cols + other_cols + att_cols)\n",
    "\n",
    "melpomene_reference_gene_info_df = pd.DataFrame(\n",
    "    rows,\n",
    "    columns=[\n",
    "        \"seqid\",\n",
    "        \"source\",\n",
    "        \"feat_type\",\n",
    "        \"start\",\n",
    "        \"end\",\n",
    "        \"score\",\n",
    "        \"strand\",\n",
    "        \"phase\",\n",
    "        \"attributes\",\n",
    "    ]\n",
    "    + other_names\n",
    "    + att_names,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "6b2d630f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>feat_type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>deletions</th>\n",
       "      <th>description</th>\n",
       "      <th>frameshifts</th>\n",
       "      <th>stable_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>translation_stable_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hmel200067o</td>\n",
       "      <td>HGC</td>\n",
       "      <td>gene</td>\n",
       "      <td>69</td>\n",
       "      <td>6264</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=HMEL005427;description=Nitric oxide synthas...</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>HMEL005427</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Nitric oxide synthase like B</td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427</td>\n",
       "      <td>NOS-B</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hmel200067o</td>\n",
       "      <td>HGC</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>69</td>\n",
       "      <td>6264</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=goslim_goa:GO:0003674,goslim_goa:GO:000...</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>HMEL005427-RA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427-RA</td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427-RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hmel200067o</td>\n",
       "      <td>HGC</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69</td>\n",
       "      <td>160</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS___0;Parent=HMEL005427-RA\\n</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>CDS___0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427-RA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hmel200067o</td>\n",
       "      <td>HGC</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2806</td>\n",
       "      <td>3016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS___0;Parent=HMEL005427-RA\\n</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>CDS___0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427-RA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hmel200067o</td>\n",
       "      <td>HGC</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3238</td>\n",
       "      <td>3407</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS___0;Parent=HMEL005427-RA\\n</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>CDS___0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HMEL005427-RA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqid source feat_type start   end score strand phase  \\\n",
       "0  Hmel200067o    HGC      gene    69  6264     .      -     .   \n",
       "1  Hmel200067o    HGC      mRNA    69  6264     .      -     .   \n",
       "2  Hmel200067o    HGC       CDS    69   160     .      -     2   \n",
       "3  Hmel200067o    HGC       CDS  2806  3016     .      -     0   \n",
       "4  Hmel200067o    HGC       CDS  3238  3407     .      -     2   \n",
       "\n",
       "                                          attributes  chromosome  ...  \\\n",
       "0  ID=HMEL005427;description=Nitric oxide synthas...           0  ...   \n",
       "1  Dbxref=goslim_goa:GO:0003674,goslim_goa:GO:000...           0  ...   \n",
       "2                  ID=CDS___0;Parent=HMEL005427-RA\\n           0  ...   \n",
       "3                  ID=CDS___0;Parent=HMEL005427-RA\\n           0  ...   \n",
       "4                  ID=CDS___0;Parent=HMEL005427-RA\\n           0  ...   \n",
       "\n",
       "              ID Name Note         Parent deletions  \\\n",
       "0     HMEL005427                                      \n",
       "1  HMEL005427-RA               HMEL005427             \n",
       "2        CDS___0            HMEL005427-RA             \n",
       "3        CDS___0            HMEL005427-RA             \n",
       "4        CDS___0            HMEL005427-RA             \n",
       "\n",
       "                    description frameshifts      stable_id symbol  \\\n",
       "0  Nitric oxide synthase like B                 HMEL005427  NOS-B   \n",
       "1                                            HMEL005427-RA          \n",
       "2                                                                   \n",
       "3                                                                   \n",
       "4                                                                   \n",
       "\n",
       "  translation_stable_id  \n",
       "0                        \n",
       "1         HMEL005427-RA  \n",
       "2                        \n",
       "3                        \n",
       "4                        \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "melpomene_reference_gene_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "074f492b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>feat_type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>deletions</th>\n",
       "      <th>description</th>\n",
       "      <th>frameshifts</th>\n",
       "      <th>stable_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>translation_stable_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4717</th>\n",
       "      <td>Hmel218003o</td>\n",
       "      <td>HGC</td>\n",
       "      <td>gene</td>\n",
       "      <td>705604</td>\n",
       "      <td>706407</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=HMEL001028;description=Optix;stable_id=HMEL...</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>HMEL001028</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Optix</td>\n",
       "      <td></td>\n",
       "      <td>HMEL001028</td>\n",
       "      <td>Optix</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqid source feat_type   start     end score strand phase  \\\n",
       "4717  Hmel218003o    HGC      gene  705604  706407     .      +     .   \n",
       "\n",
       "                                             attributes  chromosome  ...  \\\n",
       "4717  ID=HMEL001028;description=Optix;stable_id=HMEL...          18  ...   \n",
       "\n",
       "              ID Name Note Parent deletions description frameshifts  \\\n",
       "4717  HMEL001028                                  Optix               \n",
       "\n",
       "       stable_id symbol translation_stable_id  \n",
       "4717  HMEL001028  Optix                        \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info_df[(gene_info_df.chromosome == 18) & (gene_info_df.symbol == \"Optix\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00be5e20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>melpomene_start</th>\n",
       "      <th>melpomene_end</th>\n",
       "      <th>erato_start</th>\n",
       "      <th>erato_end</th>\n",
       "      <th>melpomene_length</th>\n",
       "      <th>erato_length</th>\n",
       "      <th>idy</th>\n",
       "      <th>len_r</th>\n",
       "      <th>len_q</th>\n",
       "      <th>cov_r</th>\n",
       "      <th>cov_q</th>\n",
       "      <th>melpomene_tag</th>\n",
       "      <th>erato_tag</th>\n",
       "      <th>melpomene_chromosome</th>\n",
       "      <th>melpomene_scaffold</th>\n",
       "      <th>erato_chromosome</th>\n",
       "      <th>erato_scaffold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133635</th>\n",
       "      <td>704417</td>\n",
       "      <td>707238</td>\n",
       "      <td>1249225</td>\n",
       "      <td>1252032</td>\n",
       "      <td>2822</td>\n",
       "      <td>2808</td>\n",
       "      <td>90.63</td>\n",
       "      <td>16450716</td>\n",
       "      <td>5168646</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.05</td>\n",
       "      <td>Hmel218003o</td>\n",
       "      <td>Herato1801</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       melpomene_start melpomene_end erato_start erato_end melpomene_length  \\\n",
       "133635          704417        707238     1249225   1252032             2822   \n",
       "\n",
       "       erato_length    idy     len_r    len_q cov_r cov_q melpomene_tag  \\\n",
       "133635         2808  90.63  16450716  5168646  0.02  0.05   Hmel218003o   \n",
       "\n",
       "         erato_tag  melpomene_chromosome  melpomene_scaffold  \\\n",
       "133635  Herato1801                    18                   3   \n",
       "\n",
       "        erato_chromosome  erato_scaffold  \n",
       "133635                18               1  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import numpy as np\n",
    "#\n",
    "# chrom_idx = (alignment_coords_df.melpomene_chromosome == 18)\n",
    "# gene_start = 705604\n",
    "# gene_end = 706407\n",
    "# start_in_idx = (gene_start <= alignment_coords_df.melpomene_end.astype(int)) & (gene_start >= alignment_coords_df.melpomene_start.astype(int))\n",
    "# end_in_idx = (gene_end <= alignment_coords_df.melpomene_end.astype(int)) & (gene_end >= alignment_coords_df.melpomene_start.astype(int))\n",
    "#\n",
    "# alignment_coords_df[chrom_idx & (start_in_idx | end_in_idx)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ff654a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>melpomene_chromosome</th>\n",
       "      <th>melpomene_scaffold</th>\n",
       "      <th>melpomene_start</th>\n",
       "      <th>melpomene_end</th>\n",
       "      <th>erato_chromosome</th>\n",
       "      <th>erato_scaffold</th>\n",
       "      <th>erato_start</th>\n",
       "      <th>erato_end</th>\n",
       "      <th>erato_iou_score</th>\n",
       "      <th>erato_containment_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HMEL005427</td>\n",
       "      <td>NOS-B</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>69</td>\n",
       "      <td>6264</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HMEL002216</td>\n",
       "      <td>Def</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10505178</td>\n",
       "      <td>10506149</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13904622</td>\n",
       "      <td>13904976</td>\n",
       "      <td>0.344130</td>\n",
       "      <td>0.952381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HMEL011647</td>\n",
       "      <td>TEP-B</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11230994</td>\n",
       "      <td>11243629</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>14783054</td>\n",
       "      <td>14784415</td>\n",
       "      <td>0.043981</td>\n",
       "      <td>0.430657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HMEL011434</td>\n",
       "      <td>Wnt10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>14208540</td>\n",
       "      <td>14229353</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>18432632</td>\n",
       "      <td>18457202</td>\n",
       "      <td>0.975259</td>\n",
       "      <td>0.975259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HMEL011436</td>\n",
       "      <td>Wnt6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>14260513</td>\n",
       "      <td>14295660</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>18493665</td>\n",
       "      <td>18534835</td>\n",
       "      <td>0.954252</td>\n",
       "      <td>0.954252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id gene_symbol  melpomene_chromosome  melpomene_scaffold  \\\n",
       "0  HMEL005427       NOS-B                     0                  67   \n",
       "1  HMEL002216         Def                     1                   1   \n",
       "2  HMEL011647       TEP-B                     1                   1   \n",
       "3  HMEL011434       Wnt10                     1                   1   \n",
       "4  HMEL011436        Wnt6                     1                   1   \n",
       "\n",
       "   melpomene_start  melpomene_end  erato_chromosome  erato_scaffold  \\\n",
       "0               69           6264                -1              -1   \n",
       "1         10505178       10506149                 1               1   \n",
       "2         11230994       11243629                 1               1   \n",
       "3         14208540       14229353                 1               1   \n",
       "4         14260513       14295660                 1               1   \n",
       "\n",
       "   erato_start  erato_end  erato_iou_score  erato_containment_score  \n",
       "0           -1         -1         0.000000                 0.000000  \n",
       "1     13904622   13904976         0.344130                 0.952381  \n",
       "2     14783054   14784415         0.043981                 0.430657  \n",
       "3     18432632   18457202         0.975259                 0.975259  \n",
       "4     18493665   18534835         0.954252                 0.954252  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info_df = melpomene_reference_gene_info_df[\n",
    "    melpomene_reference_gene_info_df.feat_type == \"gene\"\n",
    "]\n",
    "\n",
    "new_gene_info_rows = []\n",
    "for idx, df_row in gene_info_df.iterrows():\n",
    "    melpomene_start = int(df_row.start)\n",
    "    melpomene_end = int(df_row.end)\n",
    "    melpomene_chromosome = int(df_row.chromosome)\n",
    "    melpomene_scaffold = int(df_row.scaffold)\n",
    "    gene_id = df_row.ID\n",
    "    gene_symbol = df_row.symbol\n",
    "\n",
    "    # Find intersections\n",
    "    filtered_alignment_df = alignment_coords_df[\n",
    "        alignment_coords_df.melpomene_chromosome == melpomene_chromosome\n",
    "    ]\n",
    "\n",
    "    gene_start = int(df_row.start)\n",
    "    gene_end = int(df_row.end)\n",
    "    start_in_idx = (gene_start <= filtered_alignment_df.melpomene_end.astype(int)) & (\n",
    "        gene_start >= filtered_alignment_df.melpomene_start.astype(int)\n",
    "    )\n",
    "    end_in_idx = (gene_end <= filtered_alignment_df.melpomene_end.astype(int)) & (\n",
    "        gene_end >= filtered_alignment_df.melpomene_start.astype(int)\n",
    "    )\n",
    "\n",
    "    intersections = start_in_idx | end_in_idx\n",
    "\n",
    "    filtered_alignment_df = filtered_alignment_df[intersections]\n",
    "\n",
    "    erato_start = -1\n",
    "    erato_end = -1\n",
    "    erato_chromosome = -1\n",
    "    erato_scaffold = -1\n",
    "    iou_score = 0\n",
    "    containment_score = 0\n",
    "    if filtered_alignment_df.shape[0] > 0:\n",
    "        min_start = int(filtered_alignment_df.melpomene_start.min())\n",
    "        max_end = int(filtered_alignment_df.melpomene_end.max())\n",
    "\n",
    "        # may have to incorporate the idy column from the aligment scores in these\n",
    "        intersection = max(\n",
    "            0, min(melpomene_end, max_end) - max(melpomene_start, min_start)\n",
    "        )\n",
    "        union = max(1, max(melpomene_end, max_end) - min(melpomene_start, min_start))\n",
    "\n",
    "        if intersection == 0:\n",
    "            continue\n",
    "\n",
    "        iou_score = intersection / union\n",
    "        containment_score = intersection / max(1, max_end - min_start)\n",
    "        erato_start = int(filtered_alignment_df.erato_start.min())\n",
    "        erato_end = int(filtered_alignment_df.erato_end.max())\n",
    "\n",
    "        erato_chromosome = int(\n",
    "            filtered_alignment_df.erato_chromosome.max()\n",
    "        )  # Should be unique\n",
    "\n",
    "        erato_scaffold = int(\n",
    "            filtered_alignment_df.erato_scaffold.max()\n",
    "        )  # May not be unique...so don't count on this\n",
    "\n",
    "    new_gene_info_rows.append(\n",
    "        [\n",
    "            gene_id,\n",
    "            gene_symbol,\n",
    "            melpomene_chromosome,\n",
    "            melpomene_scaffold,\n",
    "            melpomene_start,\n",
    "            melpomene_end,\n",
    "            erato_chromosome,\n",
    "            erato_scaffold,\n",
    "            erato_start,\n",
    "            erato_end,\n",
    "            iou_score,\n",
    "            containment_score,\n",
    "        ]\n",
    "    )\n",
    "\n",
    "gene_aligment_info_df = pd.DataFrame(\n",
    "    new_gene_info_rows,\n",
    "    columns=[\n",
    "        \"gene_id\",\n",
    "        \"gene_symbol\",\n",
    "        \"melpomene_chromosome\",\n",
    "        \"melpomene_scaffold\",\n",
    "        \"melpomene_start\",\n",
    "        \"melpomene_end\",\n",
    "        \"erato_chromosome\",\n",
    "        \"erato_scaffold\",\n",
    "        \"erato_start\",\n",
    "        \"erato_end\",\n",
    "        \"erato_iou_score\",\n",
    "        \"erato_containment_score\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "gene_aligment_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "999fe1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_aligment_info_df.to_csv(DATA_DIR / \"gene_alignment.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
