{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Editing Dashboard\n",
    "\n",
    "This notebook is intended to give an interactive dashboard to view changes to the phenotype when changing the DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import base64\n",
    "from pathlib import Path\n",
    "import functools\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "from gtp.configs.loaders import load_configs\n",
    "from gtp.configs.project import GenotypeToPhenotypeConfigs\n",
    "from gtp.dataloading.path_collectors import (\n",
    "    get_post_processed_genotype_directory,\n",
    "    get_results_plot_output_directory,\n",
    "    get_results_training_metadata_directory,\n",
    ")\n",
    "from gtp.dataloading.tools import collect_chromosome_position_metadata\n",
    "from gtp.options.plot_attribution import PlotAttributionOptions\n",
    "\n",
    "config_path = Path(\"../configs/default.yaml\")\n",
    "cfgs: GenotypeToPhenotypeConfigs = load_configs(config_path)\n",
    "opts: PlotAttributionOptions = PlotAttributionOptions(\n",
    "    species=\"erato\",\n",
    "    color=\"color_3\",\n",
    "    wing=\"forewings\",\n",
    "    exp_name=\"base\",\n",
    "    verbose=True,\n",
    ")\n",
    "\n",
    "\n",
    "# Using base64 encoding and decoding\n",
    "def b64_image(image_filename):\n",
    "    with open(image_filename, \"rb\") as f:\n",
    "        image = f.read()\n",
    "    return \"data:image/png;base64,\" + base64.b64encode(image).decode(\"utf-8\")\n",
    "\n",
    "\n",
    "def get_attribution_plot():\n",
    "    plot_output_dir = get_results_plot_output_directory(cfgs.io)\n",
    "    plot_output_dir_final = (\n",
    "        plot_output_dir / opts.exp_name / f\"{opts.species}_{opts.wing}_{opts.color}\"\n",
    "    )\n",
    "    plot_output_dir_final.mkdir(exist_ok=True, parents=True)\n",
    "    img_path = plot_output_dir_final / \"manhattan_plot_Attribution.png\"\n",
    "    if not img_path.exists():\n",
    "        # Backwards compatibility from recent change. Delete this if after re run of experiments\n",
    "        img_path = plot_output_dir_final / \"manhattan_plot.png\"\n",
    "\n",
    "    return b64_image(img_path)\n",
    "\n",
    "\n",
    "def load_camids():\n",
    "    metadata_dir = get_results_training_metadata_directory(cfgs.io)\n",
    "    train_camids = np.load(metadata_dir / f\"{opts.species}_train.npy\").tolist()\n",
    "    val_camids = np.load(metadata_dir / f\"{opts.species}_val.npy\").tolist()\n",
    "    test_camids = np.load(metadata_dir / f\"{opts.species}_test.npy\").tolist()\n",
    "    camids = train_camids + val_camids + test_camids\n",
    "\n",
    "    return camids\n",
    "\n",
    "\n",
    "def load_genes():\n",
    "    genotype_folder = get_post_processed_genotype_directory(cfgs.io)\n",
    "\n",
    "    position_metadata = collect_chromosome_position_metadata(\n",
    "        genotype_folder / cfgs.experiment.genotype_scope,\n",
    "        opts.species,\n",
    "        opts.chromosome,\n",
    "    )\n",
    "\n",
    "    genes = [f\"{scaffold}:{position}\" for scaffold, position in position_metadata]\n",
    "\n",
    "    return genes[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash app running on http://127.0.0.1:8050/\n"
     ]
    }
   ],
   "source": [
    "from dash import Dash, html, dcc, callback, Output, Input, jupyter_dash\n",
    "import plotly.express as px\n",
    "import pandas as pd\n",
    "\n",
    "jupyter_dash.default_mode = \"external\"\n",
    "\n",
    "species = [\"erato\", \"melpomene\"]\n",
    "wings = [\"forewings\", \"hindwings\"]\n",
    "phenotypes = [\"color_1\", \"color_2\", \"color_3\", \"total\"]\n",
    "PCA_DIMS = 1\n",
    "pca_dim = [i for i in range(PCA_DIMS)]\n",
    "gene_edits = [\"AA\", \"Aa/aA\", \"aa\"]\n",
    "\n",
    "app = Dash()\n",
    "\n",
    "app.layout = [\n",
    "    html.H1(children=\"Gene Editing Dashboard\", style={\"textAlign\": \"center\"}),\n",
    "    dcc.Dropdown(species, opts.species, id=\"species-selection\"),\n",
    "    dcc.Dropdown(wings, opts.wing, id=\"wing-selection\"),\n",
    "    dcc.Dropdown(phenotypes, opts.color, id=\"phenotype-selection\"),\n",
    "    dcc.Dropdown(pca_dim, 0, id=\"output-dim-selection\"),\n",
    "    html.H1(id=\"custom_text\", children=\"Template\"),\n",
    "    html.Img(id=\"attribution-plot\", src=get_attribution_plot(), width=600, height=400),\n",
    "    dcc.Dropdown([], \"\", id=\"specimen-selection\"),\n",
    "    dcc.Dropdown([], \"\", id=\"gene-selection\"),\n",
    "    dcc.Dropdown(gene_edits, \"\", id=\"gene-edit-selection\"),\n",
    "]\n",
    "\n",
    "\n",
    "def create_title_str():\n",
    "    return f\"{opts.species}-{opts.wing}-{opts.color}-{0}\"\n",
    "\n",
    "\n",
    "@callback(\n",
    "    Output(\"custom_text\", \"children\"),\n",
    "    Input(\"species-selection\", \"value\"),\n",
    "    Input(\"wing-selection\", \"value\"),\n",
    "    Input(\"phenotype-selection\", \"value\"),\n",
    "    Input(\"output-dim-selection\", \"value\"),\n",
    ")\n",
    "def update_model(species, wing, phenotype, output_dim):\n",
    "    opts.species = species\n",
    "    opts.wing = wing\n",
    "    opts.color = phenotype\n",
    "    return [create_title_str()]\n",
    "\n",
    "\n",
    "@callback(\n",
    "    Output(\"attribution-plot\", \"src\"),\n",
    "    Input(\"species-selection\", \"value\"),\n",
    "    Input(\"wing-selection\", \"value\"),\n",
    "    Input(\"phenotype-selection\", \"value\"),\n",
    "    Input(\"output-dim-selection\", \"value\"),\n",
    ")\n",
    "def update_plot(species, wing, phenotype, output_dim):\n",
    "    opts.species = species\n",
    "    opts.wing = wing\n",
    "    opts.color = phenotype\n",
    "    return get_attribution_plot()\n",
    "\n",
    "\n",
    "@callback(\n",
    "    Output(\"specimen-selection\", \"options\"),\n",
    "    Output(\"specimen-selection\", \"value\"),\n",
    "    Input(\"species-selection\", \"value\"),\n",
    "    Input(\"wing-selection\", \"value\"),\n",
    "    Input(\"phenotype-selection\", \"value\"),\n",
    "    Input(\"output-dim-selection\", \"value\"),\n",
    ")\n",
    "def update_specimen_selections(species, wing, phenotype, output_dim):\n",
    "    opts.species = species\n",
    "    opts.wing = wing\n",
    "    opts.color = phenotype\n",
    "    camids = load_camids()\n",
    "    return camids, camids[0]\n",
    "\n",
    "\n",
    "@callback(\n",
    "    Output(\"gene-selection\", \"options\"),\n",
    "    Output(\"gene-selection\", \"value\"),\n",
    "    Input(\"species-selection\", \"value\"),\n",
    "    Input(\"wing-selection\", \"value\"),\n",
    "    Input(\"phenotype-selection\", \"value\"),\n",
    "    Input(\"output-dim-selection\", \"value\"),\n",
    ")\n",
    "def update_gene_selections(species, wing, phenotype, output_dim):\n",
    "    opts.species = species\n",
    "    opts.wing = wing\n",
    "    opts.color = phenotype\n",
    "    genes = load_genes()\n",
    "    return genes, genes[0]\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    app.run(debug=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
