{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "from pathlib import Path\n",
    "from io import BytesIO\n",
    "\n",
    "import torch\n",
    "from captum.attr import LRP\n",
    "\n",
    "from gtp.dataloading.tools import (\n",
    "    collect_chromosome,\n",
    "    collect_chromosome_position_metadata,\n",
    ")\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageChops\n",
    "from matplotlib import cm\n",
    "\n",
    "from gtp.configs.loaders import load_configs\n",
    "from gtp.configs.project import GenotypeToPhenotypeConfigs\n",
    "from gtp.dataloading.path_collectors import (\n",
    "    get_experiment_directory,\n",
    "    get_post_processed_genotype_directory,\n",
    "    get_post_processed_phenotype_directory,\n",
    "    get_results_training_metadata_directory,\n",
    ")\n",
    "from gtp.models.net import SoyBeanNet\n",
    "from gtp.options.process_attribution import ProcessAttributionOptions\n",
    "\n",
    "os.environ[\"CUDA_DEVICE_ORDER\"] = \"PCI_BUS_ID\"\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"1\"\n",
    "\n",
    "config_path = Path(\"../configs/default.yaml\")\n",
    "cfgs: GenotypeToPhenotypeConfigs = load_configs(config_path)\n",
    "opts: ProcessAttributionOptions = ProcessAttributionOptions(\n",
    "    drop_out_prob=0.75,\n",
    "    out_dims=1,\n",
    "    out_dims_start_idx=0,\n",
    "    insize=3,\n",
    "    hidden_dim=10,\n",
    "    species=\"erato\",\n",
    "    chromosome=18,\n",
    "    color=\"color_3\",\n",
    "    wing=\"forewings\",\n",
    "    exp_name=\"base\",\n",
    ")\n",
    "\n",
    "\n",
    "def get_camids(species):\n",
    "    metadata_dir = get_results_training_metadata_directory(cfgs.io)\n",
    "    train_camids = np.load(metadata_dir / f\"{species}_train.npy\").tolist()\n",
    "    val_camids = np.load(metadata_dir / f\"{species}_val.npy\").tolist()\n",
    "    test_camids = np.load(metadata_dir / f\"{species}_test.npy\").tolist()\n",
    "    camids = train_camids + val_camids + test_camids\n",
    "\n",
    "    return sorted(camids)\n",
    "\n",
    "\n",
    "def get_proj_matrix(species, wing, color):\n",
    "    proj_matrices_dir = Path(cfgs.io.default_root, \"dna/projection_matrices\")\n",
    "    pca_df = pd.read_csv(proj_matrices_dir / f\"{species}_{wing}_{color}.csv\")\n",
    "    pca_w = pca_df.to_numpy()\n",
    "    return pca_w\n",
    "\n",
    "\n",
    "def create_proj_img_bytes(pca_w, pca_vector):\n",
    "    proj_img_m = pca_w @ pca_vector.T\n",
    "    proj_img_m = proj_img_m.reshape(300, 300)  # Range between [-1, 1]\n",
    "    # proj_img_m += 1\n",
    "    # proj_img_m /= 2  # [0, 1]\n",
    "    proj_img_m[proj_img_m <= 0] = 0\n",
    "    proj_img_m[proj_img_m > 0] = 1\n",
    "    im = Image.fromarray(np.uint8(cm.bwr(proj_img_m) * 255))\n",
    "\n",
    "    im_bytes = BytesIO()\n",
    "    im.save(im_bytes, format=\"PNG\")\n",
    "    return im_bytes\n",
    "\n",
    "\n",
    "def get_proj_img(species, wing, color, camid):\n",
    "    phenotype_folder = get_post_processed_phenotype_directory(cfgs.io)\n",
    "    pca_df = pd.read_csv(phenotype_folder / f\"{species}_{wing}_{color}\" / \"data.csv\")\n",
    "    results = pca_df.loc[pca_df.camid == camid]\n",
    "    pca_vector = results.iloc[:1, 1:].to_numpy()\n",
    "    pca_w = get_proj_matrix(species, wing, color)\n",
    "\n",
    "    im_bytes = create_proj_img_bytes(pca_w, pca_vector)\n",
    "\n",
    "    return im_bytes\n",
    "\n",
    "\n",
    "def get_model_proj_img(species, wing, color, camid, pca_edit_vec, start_idx):\n",
    "    phenotype_folder = get_post_processed_phenotype_directory(cfgs.io)\n",
    "    pca_df = pd.read_csv(phenotype_folder / f\"{species}_{wing}_{color}\" / \"data.csv\")\n",
    "    results = pca_df.loc[pca_df.camid == camid]\n",
    "    pca_vector = results.iloc[:1, 1:].to_numpy()\n",
    "    pca_vector[0][start_idx : len(pca_edit_vec)] = pca_edit_vec\n",
    "    pca_w = get_proj_matrix(species, wing, color)\n",
    "\n",
    "    im_bytes = create_proj_img_bytes(pca_w, pca_vector)\n",
    "\n",
    "    return im_bytes\n",
    "\n",
    "\n",
    "def get_genes(species, chromosome, camid):\n",
    "    # Get gene states\n",
    "    genotype_folder = get_post_processed_genotype_directory(cfgs.io)\n",
    "    genotype_folder = genotype_folder / cfgs.experiment.genotype_scope\n",
    "    genotype_camids, genotype_data = collect_chromosome(\n",
    "        genotype_folder, species, chromosome\n",
    "    )\n",
    "    loc = np.where(genotype_camids == camid)[0][0]\n",
    "    genotype_data_short = (genotype_data[loc] * [0, 1, 2]).sum(-1)\n",
    "    string_genes = np.empty(genotype_data_short.shape[0], dtype=\"<U10\")\n",
    "    string_genes[genotype_data_short == 0] = \"aa\"  # Recessive\n",
    "    string_genes[genotype_data_short == 1] = \"Aa/aA\"  # heterogeneous\n",
    "    string_genes[genotype_data_short == 2] = \"AA\"  # Dominant\n",
    "\n",
    "    # Get Position Metadata\n",
    "    position_metadata = collect_chromosome_position_metadata(\n",
    "        genotype_folder,\n",
    "        species,\n",
    "        chromosome,\n",
    "    )\n",
    "\n",
    "    pos_strs = [f\"{x[0]}:{x[1]}\" for x in position_metadata]\n",
    "\n",
    "    assert string_genes.shape[0] == len(pos_strs), \"Error in alignment\"\n",
    "    rv = [f\"{x[0]} ({x[1]})\" for x in zip(pos_strs, string_genes)]\n",
    "    return rv, genotype_data[loc]\n",
    "\n",
    "\n",
    "def load_model(species, wing, color, chromosome, camid):\n",
    "    start = time.time()\n",
    "    gene_input = (\n",
    "        torch.tensor(get_genes(species, chromosome, camid)[1]).unsqueeze(0).unsqueeze(0)\n",
    "    )\n",
    "    num_vcfs = gene_input[0].shape[1]\n",
    "    model = SoyBeanNet(\n",
    "        window_size=num_vcfs,\n",
    "        num_out_dims=opts.out_dims,\n",
    "        insize=opts.insize,\n",
    "        hidden_dim=opts.hidden_dim,\n",
    "        drop_out_prob=opts.drop_out_prob,\n",
    "    )\n",
    "\n",
    "    experiment_dir = get_experiment_directory(\n",
    "        cfgs.io,\n",
    "        species=species,\n",
    "        wing=wing,\n",
    "        color=color,\n",
    "        chromosome=chromosome,\n",
    "        exp_name=opts.exp_name,\n",
    "    )\n",
    "\n",
    "    model.load_state_dict(torch.load(experiment_dir / \"model.pt\", weights_only=True))\n",
    "    model = model.cuda()\n",
    "    model.eval()\n",
    "\n",
    "    end = time.time()\n",
    "    print(f\"Time to load the model: {end - start}\")\n",
    "    return model\n",
    "\n",
    "\n",
    "def get_lrp_attr_single(model, x, target=0):\n",
    "    att_model = LRP(model)\n",
    "    model.zero_grad()\n",
    "    x.requires_grad = True\n",
    "    attr = att_model.attribute(x.cuda(), target=target)\n",
    "    # For LRP, this (ONE-HOT state ex. [0,0,1] attributions) should be sum.\n",
    "    # This is because the attribution scores should all add up to be the find value in the prediction, so averaging could break that.\n",
    "    attr = attr.sum(-1)\n",
    "    attr = attr[0, 0]  # Only has 1 channel, just extract it and is one batch item\n",
    "    attr = attr.detach().cpu().numpy()\n",
    "\n",
    "    return attr\n",
    "\n",
    "\n",
    "def get_top_n_peak_windows(init_attr, n=3, window_size=10_000):\n",
    "    abs_attr = np.copy(np.abs(init_attr))\n",
    "    D = len(init_attr)\n",
    "    peak_windows = []\n",
    "    for ni in range(n):\n",
    "        sorted_idx = np.argsort(abs_attr)[::-1]\n",
    "        min_idx = max(0, sorted_idx[0] - window_size)\n",
    "        max_idx = min(D - 1, sorted_idx[0] + window_size)\n",
    "        peak_windows.append([min_idx, max_idx])\n",
    "        abs_attr[min_idx:max_idx] = 0.0\n",
    "\n",
    "    return peak_windows\n",
    "\n",
    "\n",
    "test_wing = \"forewings\"\n",
    "test_species = \"erato\"\n",
    "test_color = \"color_3\"\n",
    "test_chromosome = 18\n",
    "test_camid = \"CAM016634\"\n",
    "# gene_str, gene_ml = get_genes(test_species, test_chromosome, test_camid)\n",
    "##\n",
    "# m = load_model(test_species, test_wing, test_color, test_chromosome, test_camid)\n",
    "# m = m.cuda()\n",
    "# get_lrp_attr_single(m, torch.tensor(gene_ml).unsqueeze(0).unsqueeze(0).float().cuda())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipywidgets as widgets\n",
    "import time\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "debug_text_w = widgets.Text(\n",
    "    value=\"\", placeholder=\"Type something\", description=\"Debug Out:\", disabled=False\n",
    ")\n",
    "\n",
    "species_dd = widgets.Dropdown(\n",
    "    options=[\"erato\", \"melpomene\"],\n",
    "    value=\"erato\",\n",
    "    description=\"Species\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "wing_dd = widgets.Dropdown(\n",
    "    options=[\"forewings\", \"hindwings\"],\n",
    "    value=\"forewings\",\n",
    "    description=\"Wing\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "color_dd = widgets.Dropdown(\n",
    "    options=[\"color_1\", \"color_2\", \"color_3\", \"total\"],\n",
    "    value=\"color_3\",\n",
    "    description=\"Color\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "chromosome_box = widgets.BoundedIntText(\n",
    "    value=18, min=1, max=21, step=1, description=\"Chromosome:\", disabled=False\n",
    ")\n",
    "\n",
    "init_camids = get_camids(\"erato\")\n",
    "camid_dd = widgets.Dropdown(\n",
    "    options=init_camids,\n",
    "    value=init_camids[0],\n",
    "    description=\"CAMID\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "proj_img = widgets.Image(\n",
    "    value=get_proj_img(\n",
    "        species=\"erato\", wing=\"forewings\", color=\"color_3\", camid=init_camids[0]\n",
    "    ).getvalue(),\n",
    "    format=\"png\",\n",
    ")\n",
    "\n",
    "model = load_model(\"erato\", \"forewings\", \"color_3\", 18, init_camids[0])\n",
    "\n",
    "init_gene_strs, init_gene_ml = get_genes(\"erato\", 18, init_camids[0])\n",
    "gene_input = torch.tensor(init_gene_ml).unsqueeze(0).unsqueeze(0).float().cuda()\n",
    "init_attr = get_lrp_attr_single(model, gene_input)\n",
    "attr_order = np.argsort(np.abs(init_attr))[::-1]\n",
    "\n",
    "start = time.time()\n",
    "output = model(gene_input)[0].detach().cpu().numpy()\n",
    "last_output = output\n",
    "end = time.time()\n",
    "print(f\"Time to use the model: {end - start}\")\n",
    "\n",
    "\n",
    "model_proj_bytes = get_model_proj_img(\n",
    "    \"erato\", \"forewings\", \"color_3\", init_camids[0], output, opts.out_dims_start_idx\n",
    ").getvalue()\n",
    "\n",
    "model_proj_img = widgets.Image(\n",
    "    value=model_proj_bytes,\n",
    "    format=\"png\",\n",
    ")\n",
    "\n",
    "edit_model_proj_img = widgets.Image(\n",
    "    value=model_proj_bytes,\n",
    "    format=\"png\",\n",
    ")\n",
    "\n",
    "zero_img = Image.fromarray(np.zeros((300, 300)).astype(np.uint8))\n",
    "zero_img_bytes = BytesIO()\n",
    "zero_img.save(zero_img_bytes, format=\"PNG\")\n",
    "\n",
    "diff_edit_model_proj_img = widgets.Image(\n",
    "    value=zero_img_bytes.getvalue(),\n",
    "    format=\"png\",\n",
    ")\n",
    "\n",
    "\n",
    "def get_base_values():\n",
    "    return species_dd.value, wing_dd.value, color_dd.value, chromosome_box.value\n",
    "\n",
    "\n",
    "def update_camids(*args):\n",
    "    camids = get_camids(species_dd.value)\n",
    "    camid_dd.options = camids\n",
    "    camid_dd.value = camids[0]\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(9, 3))\n",
    "ax.scatter(np.arange(len(init_attr)), init_attr)\n",
    "peak_windows = get_top_n_peak_windows(init_attr)\n",
    "for pw in peak_windows:\n",
    "    ax.axvline(pw[0], color=\"orange\")\n",
    "    ax.axvline(pw[1], color=\"orange\")\n",
    "buf = BytesIO()\n",
    "fig.savefig(buf, format=\"png\")\n",
    "\n",
    "attribution_plot = widgets.Image(\n",
    "    value=buf.getvalue(),\n",
    "    format=\"png\",\n",
    ")\n",
    "\n",
    "\n",
    "def update_proj_img(*args):\n",
    "    attribution_plot.layout.visibility = \"hidden\"\n",
    "    proj_img.layout.visibility = \"hidden\"\n",
    "    model_proj_img.layout.visibility = \"hidden\"\n",
    "    species, wing, color, chromosome = get_base_values()\n",
    "    camid = camid_dd.value\n",
    "\n",
    "    proj_bytes = get_proj_img(\n",
    "        species=species, wing=wing, color=color, camid=camid\n",
    "    ).getvalue()\n",
    "    proj_img.value = proj_bytes\n",
    "    proj_img.layout.visibility = \"visible\"\n",
    "\n",
    "    model = load_model(species, wing, color, chromosome, camid)\n",
    "    gene_input = (\n",
    "        torch.tensor(get_genes(species, chromosome, camid)[1])\n",
    "        .unsqueeze(0)\n",
    "        .unsqueeze(0)\n",
    "        .float()\n",
    "    )\n",
    "\n",
    "    output = model(gene_input.cuda())[0].detach().cpu().numpy()\n",
    "    last_output = output\n",
    "    model_proj_bytes = get_model_proj_img(\n",
    "        species, wing, color, camid, output, opts.out_dims_start_idx\n",
    "    ).getvalue()\n",
    "    model_proj_img.value = model_proj_bytes\n",
    "    model_proj_img.layout.visibility = \"visible\"\n",
    "\n",
    "    edit_model_proj_img.value = model_proj_bytes\n",
    "    diff_edit_model_proj_img.value = zero_img_bytes.getvalue()\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(9, 3))\n",
    "    ax.scatter(np.arange(len(init_attr)), init_attr)\n",
    "    peak_windows = get_top_n_peak_windows(init_attr, window_size=window_size_dd.value)\n",
    "    for pw in peak_windows:\n",
    "        ax.axvline(pw[0], color=\"orange\")\n",
    "        ax.axvline(pw[1], color=\"orange\")\n",
    "    buf = BytesIO()\n",
    "    fig.savefig(buf, format=\"png\")\n",
    "\n",
    "    attribution_plot.value = buf.getvalue()\n",
    "    attribution_plot.layout.visibility = \"visible\"\n",
    "\n",
    "\n",
    "species_dd.observe(update_camids, \"value\")\n",
    "species_dd.observe(update_proj_img, \"value\")\n",
    "wing_dd.observe(update_proj_img, \"value\")\n",
    "color_dd.observe(update_proj_img, \"value\")\n",
    "camid_dd.observe(update_proj_img, \"value\")\n",
    "\n",
    "display_genes = [\n",
    "    f\"{init_gene_strs[idx]} {round(init_attr[idx], 2)}\" for idx in attr_order[:50]\n",
    "]\n",
    "\n",
    "gene_selection_dd = widgets.Dropdown(\n",
    "    options=display_genes,\n",
    "    value=display_genes[0],\n",
    "    description=\"Gene\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "\n",
    "def update_gene_selection(*args):\n",
    "    species, wing, color, chromosome = get_base_values()\n",
    "    genes_str, genes_ml = get_genes(species, chromosome, camid_dd.value)\n",
    "    init_attr = get_lrp_attr_single(model, gene_input)\n",
    "    attr_order = np.argsort(np.abs(init_attr))[::-1]\n",
    "    display_genes = [\n",
    "        f\"{genes_str[idx]} {round(init_attr[idx], 2)}\" for idx in attr_order[:50]\n",
    "    ]\n",
    "    gene_selection_dd.options = display_genes[:50]\n",
    "    gene_selection_dd.value = display_genes[0]\n",
    "\n",
    "\n",
    "species_dd.observe(update_gene_selection, \"value\")\n",
    "chromosome_box.observe(update_gene_selection, \"value\")\n",
    "camid_dd.observe(update_gene_selection, \"value\")\n",
    "\n",
    "init_gene_str = display_genes[0].split(\" \")[1][1:-1]\n",
    "gene_edit_selection_dd = widgets.Dropdown(\n",
    "    options=[\"AA\", \"Aa/aA\", \"aa\"],\n",
    "    value=init_gene_str,\n",
    "    description=\"Gene Edit:\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "window_size_dd = widgets.Dropdown(\n",
    "    options=list(range(0, 100001, 10000)),\n",
    "    value=0,\n",
    "    description=\"Window Size\",\n",
    "    disable=False,\n",
    ")\n",
    "\n",
    "diff_img = None\n",
    "\n",
    "\n",
    "def update_model_edit_proj_img(*args):\n",
    "    edit_model_proj_img.layout.visibility = \"hidden\"\n",
    "    diff_edit_model_proj_img.layout.visibility = \"hidden\"\n",
    "    species, wing, color, chromosome = get_base_values()\n",
    "    camid = camid_dd.value\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(9, 3))\n",
    "    ax.scatter(np.arange(len(init_attr)), init_attr)\n",
    "    peak_windows = get_top_n_peak_windows(init_attr, window_size=window_size_dd.value)\n",
    "    for pw in peak_windows:\n",
    "        ax.axvline(pw[0], color=\"orange\")\n",
    "        ax.axvline(pw[1], color=\"orange\")\n",
    "    buf = BytesIO()\n",
    "    fig.savefig(buf, format=\"png\")\n",
    "    attribution_plot.value = buf.getvalue()\n",
    "\n",
    "    model = load_model(species, wing, color, chromosome, camid)\n",
    "    gene_input = (\n",
    "        torch.tensor(get_genes(species, chromosome, camid)[1])\n",
    "        .unsqueeze(0)\n",
    "        .unsqueeze(0)\n",
    "        .float()\n",
    "    )\n",
    "\n",
    "    edit_vec = None\n",
    "    edit_gene_str = gene_edit_selection_dd.value\n",
    "    if edit_gene_str == \"AA\":\n",
    "        edit_vec = torch.tensor([0, 0, 1])\n",
    "    elif edit_gene_str == \"aa\":\n",
    "        edit_vec = torch.tensor([1, 0, 0])\n",
    "    else:\n",
    "        edit_vec = torch.tensor([0, 1, 0])\n",
    "\n",
    "    ws = window_size_dd.value\n",
    "    if ws == 0:\n",
    "        ws = 1\n",
    "    else:\n",
    "        edit_vec = edit_vec.unsqueeze(0).repeat(ws * 2, 1)\n",
    "\n",
    "    loc = attr_order[gene_selection_dd.options.index(gene_selection_dd.value)]\n",
    "\n",
    "    if ws == 1:\n",
    "        gene_input[0][0][loc] = edit_vec\n",
    "    else:\n",
    "        gene_input[0][0][loc - ws : loc + ws] = edit_vec\n",
    "\n",
    "    output = model(gene_input.cuda())[0].detach().cpu().numpy()\n",
    "    print(output)\n",
    "    debug_text_w.value = f\"{output}\"\n",
    "    print(f\"Output diff: {last_output - output}\")\n",
    "    edit_model_proj_bytes = get_model_proj_img(\n",
    "        species, wing, color, camid, output, opts.out_dims_start_idx\n",
    "    ).getvalue()\n",
    "    edit_model_proj_img.value = edit_model_proj_bytes\n",
    "\n",
    "    model_img = Image.open(BytesIO(model_proj_img.value))\n",
    "    edit_model_img = Image.open(BytesIO(edit_model_proj_bytes))\n",
    "\n",
    "    diff_img = (np.array(model_img) - np.array(edit_model_img)).astype(np.float64)\n",
    "    print(diff_img[:10, :10, 0])\n",
    "    diff_img = diff_img.sum(-1)\n",
    "    diff_img -= diff_img.min()\n",
    "    diff_img /= diff_img.max()\n",
    "\n",
    "    diff_img = Image.fromarray(\n",
    "        np.uint8(\n",
    "            cm.bwr(\n",
    "                diff_img,\n",
    "            )\n",
    "            * 255\n",
    "        )\n",
    "    )\n",
    "    diff_bytes = BytesIO()\n",
    "    diff_img.save(diff_bytes, format=\"png\")\n",
    "    diff_edit_model_proj_img.value = diff_bytes.getvalue()\n",
    "\n",
    "    diff_edit_model_proj_img.layout.visibility = \"visible\"\n",
    "    edit_model_proj_img.layout.visibility = \"visible\"\n",
    "\n",
    "\n",
    "gene_edit_selection_dd.observe(update_model_edit_proj_img, \"value\")\n",
    "window_size_dd.observe(update_model_edit_proj_img, \"value\")\n",
    "\n",
    "grid = widgets.GridspecLayout(4, 2)\n",
    "grid[0, 0] = species_dd\n",
    "grid[0, 1] = wing_dd\n",
    "grid[1, 0] = color_dd\n",
    "grid[1, 1] = chromosome_box\n",
    "grid[2, 0] = camid_dd\n",
    "grid[2, 1] = gene_selection_dd\n",
    "grid[3, 0] = gene_edit_selection_dd\n",
    "grid[3, 1] = window_size_dd\n",
    "\n",
    "caption_grid = widgets.GridspecLayout(1, 4)\n",
    "caption_grid[0, 0] = widgets.HTML(value=\"<b>Original Projection</b>\")\n",
    "caption_grid[0, 1] = widgets.HTML(value=\"<b>Model Projection</b>\")\n",
    "caption_grid[0, 2] = widgets.HTML(value=\"<b>Edited Projection</b>\")\n",
    "caption_grid[0, 3] = widgets.HTML(value=\"<b>Difference</b>\")\n",
    "\n",
    "img_grid = widgets.GridspecLayout(1, 4)\n",
    "img_grid[0, 0] = proj_img\n",
    "img_grid[0, 1] = model_proj_img\n",
    "img_grid[0, 2] = edit_model_proj_img\n",
    "img_grid[0, 3] = diff_edit_model_proj_img\n",
    "\n",
    "display(grid)\n",
    "display(attribution_plot)\n",
    "display(caption_grid)\n",
    "display(img_grid)\n",
    "display(debug_text_w)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
