{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "import types\n",
    "\n",
    "ARGS = types.SimpleNamespace()\n",
    "ARGS.species = \"erato\"\n",
    "ARGS.color = \"color_3\"\n",
    "ARGS.wing = \"forewings\"\n",
    "ARGS.genome_folder = \"/local/scratch/carlyn.1/dna/vcfs/processed/genome\"\n",
    "ARGS.phenotype_folder = \"/local/scratch/carlyn.1/dna/colors/processed\"\n",
    "#ARGS.genome_folder = \"/local/scratch/david/geno-pheno-data/dna/processed/genome\"\n",
    "#ARGS.phenotype_folder = \"/local/scratch/david/geno-pheno-data/colors/processed/\"\n",
    "ARGS.out_dims = 1\n",
    "ARGS.out_dims_start_idx = 0\n",
    "ARGS.split_data_folder = \"/home/carlyn.1/dna-trait-analysis/data\"\n",
    "ARGS.top_k_chromosome_training_path = \"/home/carlyn.1/dna-trait-analysis/plot_results/pvalue_erato_forewings_color_3/top_k_snps_erato_forewings_color_3.npy\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "import copy\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "from tqdm import tqdm\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "from gtp.dataloading.tools import (\n",
    "    load_chromosome_data,\n",
    "    split_data_by_file,\n",
    ")\n",
    "\n",
    "def load_data(args):\n",
    "    camids_aligned, genotype_data_aligned, phenotype_data_aligned = (\n",
    "        load_chromosome_data(\n",
    "            args.genome_folder,\n",
    "            args.phenotype_folder,\n",
    "            args.species,\n",
    "            args.wing,\n",
    "            args.color,\n",
    "            args.chromosome,\n",
    "        )\n",
    "    )\n",
    "\n",
    "    phenotype_data_aligned = phenotype_data_aligned[\n",
    "        :, args.out_dims_start_idx : args.out_dims_start_idx + args.out_dims\n",
    "    ]\n",
    "\n",
    "    train_split, val_split, test_split = split_data_by_file(\n",
    "        genotype_data_aligned,\n",
    "        phenotype_data_aligned,\n",
    "        camids_aligned,\n",
    "        args.split_data_folder,\n",
    "        args.species,\n",
    "    )\n",
    "\n",
    "    return train_split, val_split, test_split\n",
    "\n",
    "def load_one(args, chromosome, snp_idx, idx):\n",
    "    cur_args = copy.deepcopy(args)\n",
    "    cur_args.chromosome = chromosome\n",
    "    train_data, val_data, test_data = load_data(cur_args)\n",
    "    snp_idx = np.sort(snp_idx).astype(np.int64)\n",
    "    train_data[0] = train_data[0][:, snp_idx]\n",
    "    val_data[0] = val_data[0][:, snp_idx]\n",
    "    test_data[0] = test_data[0][:, snp_idx]\n",
    "    return (\n",
    "        idx,\n",
    "        train_data,\n",
    "        val_data,\n",
    "        test_data,\n",
    "        snp_idx\n",
    "    )\n",
    "\n",
    "futures = []\n",
    "pool = ThreadPoolExecutor(4)\n",
    "final_train_data = None\n",
    "final_val_data = None\n",
    "final_test_data = None\n",
    "final_snps = []\n",
    "data = np.load(ARGS.top_k_chromosome_training_path, allow_pickle=True)\n",
    "test_snp_selections = data.item()[\"test\"]\n",
    "for idx, snp_idx in tqdm(enumerate(test_snp_selections), desc=\"loading top snps from chromosome\"):\n",
    "    future = pool.submit(load_one, ARGS, idx + 1, snp_idx, idx)\n",
    "    futures.append(future)\n",
    "\n",
    "total = 0\n",
    "all_data = []\n",
    "for future in as_completed(futures):\n",
    "    proc_idx, train_data, val_data, test_data, snp_idx,  = future.result()\n",
    "    all_data.append([proc_idx, train_data, val_data, test_data, snp_idx])\n",
    "    total += 1\n",
    "    print(f\"Completed loading on chromosome: {proc_idx+1}: ({total}/21)\")\n",
    "\n",
    "for proc_idx, train_data, val_data, test_data, snp_idx in sorted(all_data, key=lambda x: int(x[0])):\n",
    "    print(proc_idx)\n",
    "    final_snps.extend([f\"C_{proc_idx+1}_SNP_{si}\" for si in snp_idx])\n",
    "    if final_train_data is None:\n",
    "        final_train_data = train_data\n",
    "        final_val_data = val_data\n",
    "        final_test_data = test_data\n",
    "    else:\n",
    "        final_train_data[0] = np.concatenate(\n",
    "            (final_train_data[0], train_data[0]), axis=1\n",
    "        )\n",
    "        final_val_data[0] = np.concatenate(\n",
    "            (final_val_data[0], val_data[0]), axis=1\n",
    "        )\n",
    "        final_test_data[0] = np.concatenate(\n",
    "            (final_test_data[0], test_data[0]), axis=1\n",
    "        )\n",
    "        assert (train_data[1] == final_train_data[1]).all()\n",
    "        assert (val_data[1] == final_val_data[1]).all()\n",
    "        assert (test_data[1] == final_test_data[1]).all()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
