{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Heliconius Gene Editing Dashboard\n",
    "\n",
    "This dashboard acts as a tool to inspect perturbations on genotype-to-phenotype models.\n",
    "\n",
    "Upon editing / changing genes, the resulting effect on the phenotype will be visualized."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initial Configurations\n",
    "Edit configurations below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "from dataclasses import dataclass\n",
    "\n",
    "import torch\n",
    "\n",
    "from gtp.configs.loaders import load_configs\n",
    "from gtp.configs.project import GenotypeToPhenotypeConfigs\n",
    "from gtp.dataloading.path_collectors import (\n",
    "    get_experiment_directory,\n",
    "    get_post_processed_genotype_directory,\n",
    ")\n",
    "from gtp.dataloading.tools import collect_chromosome\n",
    "from gtp.models.net import SoyBeanNet\n",
    "from gtp.options.process_attribution import ProcessAttributionOptions\n",
    "\n",
    "# SPECIFY GPU\n",
    "os.environ[\"CUDA_DEVICE_ORDER\"] = \"PCI_BUS_ID\"\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"1\"\n",
    "\n",
    "\n",
    "class GeneEditingDashboardState:\n",
    "    def __init__(\n",
    "        self,\n",
    "        wing: str,\n",
    "        species: str,\n",
    "        color: str,\n",
    "        chromosome: int,\n",
    "        exp_name: str,\n",
    "        config_path: str,\n",
    "    ):\n",
    "        self.wing = wing\n",
    "        self.species = species\n",
    "        self.color = color\n",
    "        self.chromosome = chromosome\n",
    "        self.exp_name = exp_name\n",
    "        self.configs: GenotypeToPhenotypeConfigs = load_configs(config_path)\n",
    "        self.model = None\n",
    "        self.camids = None\n",
    "        self.data = None\n",
    "\n",
    "        self.load_data_and_model()\n",
    "\n",
    "    def load_data_and_model(self):\n",
    "        opts: ProcessAttributionOptions = ProcessAttributionOptions(\n",
    "            drop_out_prob=0.75,\n",
    "            out_dims=1,\n",
    "            out_dims_start_idx=0,\n",
    "            insize=3,\n",
    "            hidden_dim=10,\n",
    "            species=self.species,\n",
    "            chromosome=self.chromosome,\n",
    "            color=self.color,\n",
    "            wing=self.wing,\n",
    "            exp_name=self.exp_name,\n",
    "        )\n",
    "\n",
    "        processed_genotype_dir = (\n",
    "            get_post_processed_genotype_directory(self.configs.io)\n",
    "            / self.configs.experiment.genotype_scope\n",
    "        )\n",
    "\n",
    "        # Collect genotype data\n",
    "        self.camids, self.data = collect_chromosome(\n",
    "            processed_genotype_dir, self.species, self.chromosome\n",
    "        )\n",
    "\n",
    "        self.model = SoyBeanNet(\n",
    "            window_size=self.data.shape[1],\n",
    "            num_out_dims=opts.out_dims,\n",
    "            insize=opts.insize,\n",
    "            hidden_dim=opts.hidden_dim,\n",
    "            drop_out_prob=opts.drop_out_prob,\n",
    "        )\n",
    "\n",
    "        experiment_dir = get_experiment_directory(\n",
    "            self.configs.io,\n",
    "            species=self.species,\n",
    "            wing=self.wing,\n",
    "            color=self.color,\n",
    "            chromosome=self.chromosome,\n",
    "            exp_name=self.exp_name,\n",
    "        )\n",
    "\n",
    "        self.model.load_state_dict(\n",
    "            torch.load(experiment_dir / \"model.pt\", weights_only=True)\n",
    "        )\n",
    "        self.model = self.model.cuda()\n",
    "        self.model.eval()\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "EDIT HERE\n",
    "\"\"\"\n",
    "gene_editing_dashboard_state = GeneEditingDashboardState(\n",
    "    wing=\"forewings\",\n",
    "    species=\"erato\",\n",
    "    color=\"color_3\",\n",
    "    chromosome=18,\n",
    "    exp_name=\"base\",\n",
    "    config_path=Path(\"../configs/default.yaml\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dashboard\n",
    "Explore the model below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipywidgets as widgets\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from copy import copy\n",
    "\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import VBox\n",
    "\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "import numpy as np\n",
    "from captum.attr import LRP\n",
    "\n",
    "\n",
    "def get_model_input(camid):\n",
    "    input_idx = np.where(gene_editing_dashboard_state.camids == camid)[0][0]\n",
    "    x = gene_editing_dashboard_state.data[input_idx]\n",
    "    x = torch.tensor(x).unsqueeze(0).unsqueeze(0).float().cuda()\n",
    "    return x\n",
    "\n",
    "\n",
    "def get_model_output(input_x):\n",
    "    with torch.no_grad():\n",
    "        return gene_editing_dashboard_state.model(input_x)[0]\n",
    "\n",
    "\n",
    "def get_attr(camid, target=0):\n",
    "    att_model = LRP(gene_editing_dashboard_state.model)\n",
    "    gene_editing_dashboard_state.model.zero_grad()\n",
    "    x = get_model_input(camid)\n",
    "    x.requires_grad = True\n",
    "    attr = att_model.attribute(x, target=target)\n",
    "    # For LRP, this (ONE-HOT state ex. [0,0,1] attributions) should be sum.\n",
    "    # This is because the attribution scores should all add up to be the find value in the prediction, so averaging could break that.\n",
    "    attr = attr.sum(-1)\n",
    "    attr = attr[0, 0]  # Only has 1 channel, just extract it and is one batch item\n",
    "    attr = attr.detach().cpu().numpy()\n",
    "\n",
    "    return attr\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class CurrentGeneState:\n",
    "    text: str\n",
    "    index: int\n",
    "\n",
    "\n",
    "current_gene_state = CurrentGeneState(text=\"\", index=0)\n",
    "\n",
    "camid_dropdown_widget = widgets.Dropdown(\n",
    "    options=gene_editing_dashboard_state.camids,\n",
    "    value=gene_editing_dashboard_state.camids[0],\n",
    "    description=\"CAMID\",\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "current_gene_text_widget = widgets.Text(\n",
    "    value=current_gene_state.text,\n",
    "    placeholder=\"Gene\",\n",
    "    description=\"Current Gene\",\n",
    "    disabled=True,\n",
    "    layout=widgets.Layout(width=\"100%\"),\n",
    ")\n",
    "\n",
    "\n",
    "class ManhattanPlotInteractive:\n",
    "    def __init__(self):\n",
    "        self.display_box = VBox([])\n",
    "        self.update_data()\n",
    "\n",
    "    def update_data(self, *args):\n",
    "        self.camid = camid_dropdown_widget.value\n",
    "        self.Y = get_attr(self.camid)\n",
    "        self.X = np.arange(len(self.Y))\n",
    "        N = 1_000\n",
    "        top_attr_idx = np.argsort(np.abs(self.Y))[::-1]\n",
    "        self.top_n_idx = top_attr_idx[:N]\n",
    "        self.top_X = self.X[self.top_n_idx]\n",
    "        self.top_Y = self.Y[self.top_n_idx]\n",
    "\n",
    "        self.update_plot()\n",
    "\n",
    "    def update_plot(self, *args):\n",
    "        self.figure_widget = go.FigureWidget(\n",
    "            [go.Scatter(x=self.top_X, y=self.top_Y, mode=\"markers\")]\n",
    "        )\n",
    "\n",
    "        # Set Initial Colors\n",
    "        self.init_colors = [\"#0000aa\"] * len(self.top_X)\n",
    "        scatter_plot = self.get_scatter_plot()\n",
    "        scatter_plot.marker.color = self.init_colors\n",
    "\n",
    "        # Register on-click events\n",
    "        scatter_plot.on_click(self.handle_click)\n",
    "\n",
    "        self.display_box.children = [self.figure_widget]\n",
    "\n",
    "    def get_scatter_plot(self):\n",
    "        return self.figure_widget.data[0]\n",
    "\n",
    "    def handle_click(self, trace, points, state):\n",
    "        print(points)\n",
    "        idx = points.point_inds[0]\n",
    "        x = points.xs[0]\n",
    "        y = points.ys[0]\n",
    "\n",
    "        current_gene_state.index = x\n",
    "        input_x = get_model_input(camid_dropdown_widget.value)\n",
    "        vec_state = input_x[0][0][x]\n",
    "        gene_str = \"\"\n",
    "        if vec_state[0].item() == 1:\n",
    "            gene_str = \"aa\"\n",
    "        elif vec_state[1].item() == 1:\n",
    "            gene_str = \"aA/Aa\"\n",
    "        if vec_state[2].item() == 1:\n",
    "            gene_str = \"AA\"\n",
    "\n",
    "        current_gene_state.text = f\"({gene_str}) | Attribution: ({y}) | Position: ({x})\"\n",
    "        current_gene_text_widget.value = current_gene_state.text\n",
    "\n",
    "        scatter_plot = self.get_scatter_plot()\n",
    "        colors = copy(self.init_colors)\n",
    "        colors[idx] = \"#ff0000\"\n",
    "        with self.figure_widget.batch_update():\n",
    "            scatter_plot.marker.color = colors\n",
    "\n",
    "\n",
    "manhattan_plot = ManhattanPlotInteractive()\n",
    "\n",
    "camid_dropdown_widget.observe(manhattan_plot.update_data, \"value\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ipywidgets import HBox\n",
    "\n",
    "edit_btns = widgets.RadioButtons(\n",
    "    options=[\"aa\", \"aA/Aa\", \"AA\", \"zero-out\"],\n",
    ")\n",
    "edit_lbl = widgets.Label(value=\"Gene Edit\")\n",
    "gene_edit_btns = HBox([edit_lbl, edit_btns])\n",
    "\n",
    "window_size_dd = widgets.Dropdown(\n",
    "    options=list(range(0, 100001, 10000)),\n",
    "    value=0,\n",
    "    description=\"Window Size\",\n",
    "    disable=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import BytesIO\n",
    "from ipywidgets import HBox\n",
    "import pandas as pd\n",
    "\n",
    "from PIL import Image\n",
    "from matplotlib import cm\n",
    "\n",
    "from gtp.dataloading.path_collectors import get_post_processed_phenotype_directory\n",
    "\n",
    "\n",
    "def get_proj_matrix():\n",
    "    cfgs = gene_editing_dashboard_state.configs\n",
    "    species = gene_editing_dashboard_state.species\n",
    "    wing = gene_editing_dashboard_state.wing\n",
    "    color = gene_editing_dashboard_state.color\n",
    "    proj_matrices_dir = Path(cfgs.io.default_root, \"dna/projection_matrices\")\n",
    "    pca_df = pd.read_csv(proj_matrices_dir / f\"{species}_{wing}_{color}.csv\")\n",
    "    pca_w = pca_df.to_numpy()\n",
    "    return pca_w\n",
    "\n",
    "\n",
    "def create_proj_img_bytes(pca_w, pca_vector):\n",
    "    proj_img_m = pca_w @ pca_vector.T\n",
    "    proj_img_m = proj_img_m.reshape(300, 300)  # Range between [-1, 1]\n",
    "    # proj_img_m += 1\n",
    "    # proj_img_m /= 2  # [0, 1]\n",
    "    proj_img_m[proj_img_m <= 0] = 0\n",
    "    proj_img_m[proj_img_m > 0] = 1\n",
    "    im = Image.fromarray(np.uint8(cm.bwr(proj_img_m) * 255))\n",
    "\n",
    "    im_bytes = BytesIO()\n",
    "    im.save(im_bytes, format=\"PNG\")\n",
    "    return im_bytes\n",
    "\n",
    "\n",
    "def get_org_pca_vector(camid):\n",
    "    cfgs = gene_editing_dashboard_state.configs\n",
    "    species = gene_editing_dashboard_state.species\n",
    "    wing = gene_editing_dashboard_state.wing\n",
    "    color = gene_editing_dashboard_state.color\n",
    "    phenotype_folder = get_post_processed_phenotype_directory(cfgs.io)\n",
    "    pca_df = pd.read_csv(phenotype_folder / f\"{species}_{wing}_{color}\" / \"data.csv\")\n",
    "    results = pca_df.loc[pca_df.camid == camid]\n",
    "    pca_vector = results.iloc[:1, 1:].to_numpy()\n",
    "    return pca_vector\n",
    "\n",
    "\n",
    "def get_proj_img(pca_vector):\n",
    "    pca_w = get_proj_matrix()\n",
    "    im_bytes = create_proj_img_bytes(pca_w, pca_vector)\n",
    "\n",
    "    return im_bytes\n",
    "\n",
    "\n",
    "class PCAProjectionView:\n",
    "    def __init__(self):\n",
    "        self.display_box = HBox([])\n",
    "        self.update_data()\n",
    "\n",
    "    def _update_data(self):\n",
    "        camid = camid_dropdown_widget.value\n",
    "        pca_output = get_org_pca_vector(camid)\n",
    "        self.proj_img_bytes = get_proj_img(pca_output)\n",
    "\n",
    "    def update_data(self, *args):\n",
    "        self._update_data()\n",
    "        self.update_plot()\n",
    "\n",
    "    def update_plot(self):\n",
    "        data_list = [round(x, 2) for x in self.pca_output.tolist()]\n",
    "        self.figure_widget = go.FigureWidget(\n",
    "            [\n",
    "                go.Table(\n",
    "                    header=dict(values=[\"PCA\"]),\n",
    "                    cells=dict(values=[data_list]),\n",
    "                )\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        self.proj_img = widgets.Image(\n",
    "            value=self.proj_img_bytes.getvalue(),\n",
    "            format=\"png\",\n",
    "        )\n",
    "\n",
    "        self.display_box.children = [self.figure_widget, self.proj_img]\n",
    "\n",
    "\n",
    "class OrgPCAProjectionView(PCAProjectionView):\n",
    "    def _update_data(self):\n",
    "        camid = camid_dropdown_widget.value\n",
    "        pca_output = get_org_pca_vector(camid)\n",
    "        print(pca_output.shape)\n",
    "        self.proj_img_bytes = get_proj_img(pca_output)\n",
    "        self.pca_output = pca_output[0]\n",
    "\n",
    "\n",
    "class ModelPCAProjectionView(PCAProjectionView):\n",
    "    def _get_model_input(self, camid):\n",
    "        input_x = get_model_input(camid)\n",
    "        return input_x\n",
    "\n",
    "    def _update_data(self):\n",
    "        camid = camid_dropdown_widget.value\n",
    "        input_x = self._get_model_input(camid)\n",
    "        output = get_model_output(input_x)\n",
    "        pca_output = output.detach().cpu().numpy()\n",
    "        D = pca_output.shape[0]\n",
    "        pca_vector = get_org_pca_vector(camid)\n",
    "        pca_vector[0, :D] = pca_output\n",
    "        # TODO replace pca_output into vector and pass\n",
    "        self.proj_img_bytes = get_proj_img(pca_vector)\n",
    "        self.pca_output = pca_vector[0]\n",
    "\n",
    "\n",
    "class ModelEditPCAProjectionView(ModelPCAProjectionView):\n",
    "    def _get_model_input(self, camid):\n",
    "        input_x = super()._get_model_input(camid)\n",
    "        loc = current_gene_state.index\n",
    "        edit_gene_str = edit_btns.value\n",
    "        if edit_gene_str == \"AA\":\n",
    "            edit_vec = torch.tensor([0, 0, 1])\n",
    "        elif edit_gene_str == \"aa\":\n",
    "            edit_vec = torch.tensor([1, 0, 0])\n",
    "        elif edit_gene_str == \"aA/Aa\":\n",
    "            edit_vec = torch.tensor([0, 1, 0])\n",
    "        else:\n",
    "            edit_vec = torch.tensor([0, 0, 0])\n",
    "\n",
    "        ws = window_size_dd.value\n",
    "        if ws != 0:\n",
    "            min_idx = max(0, loc - ws)\n",
    "            max_idx = min(input_x.shape[2], loc + ws)\n",
    "            edit_vec = edit_vec.unsqueeze(0).repeat(max_idx - min_idx, 1)\n",
    "\n",
    "        loc = 0  # TODO: attr_order[gene_selection_dd.options.index(gene_selection_dd.value)]\n",
    "\n",
    "        if ws == 0:\n",
    "            input_x[0][0][loc] = edit_vec\n",
    "        else:\n",
    "            input_x[0][0][min_idx:max_idx] = edit_vec\n",
    "\n",
    "        return input_x\n",
    "\n",
    "\n",
    "class ModelEditDiffPCAProjectionView(PCAProjectionView):\n",
    "    def __init__(self, model_pca_proj_view, model_edit_pca_proj_view):\n",
    "        self.model_pca_proj_view: PCAProjectionView = model_pca_proj_view\n",
    "        self.model_edit_pca_proj_view: PCAProjectionView = model_edit_pca_proj_view\n",
    "        super().__init__()\n",
    "\n",
    "    def _update_data(self):\n",
    "        model_pca = self.model_pca_proj_view.pca_output\n",
    "        model_edit_pca = self.model_edit_pca_proj_view.pca_output\n",
    "        self.pca_output = model_edit_pca - model_pca\n",
    "\n",
    "        model_img = Image.open(self.model_pca_proj_view.proj_img_bytes)\n",
    "        edit_model_img = Image.open(self.model_edit_pca_proj_view.proj_img_bytes)\n",
    "\n",
    "        diff_img = (np.array(model_img) - np.array(edit_model_img)).astype(np.float64)\n",
    "        diff_img = diff_img.sum(-1)\n",
    "        diff_img -= diff_img.min()\n",
    "        if diff_img.max() != 0:\n",
    "            diff_img /= diff_img.max()\n",
    "\n",
    "        diff_img = Image.fromarray(\n",
    "            np.uint8(\n",
    "                cm.bwr(\n",
    "                    diff_img,\n",
    "                )\n",
    "                * 255\n",
    "            )\n",
    "        )\n",
    "        self.proj_img_bytes = BytesIO()\n",
    "        diff_img.save(self.proj_img_bytes, format=\"png\")\n",
    "\n",
    "\n",
    "original_pca_projection = OrgPCAProjectionView()\n",
    "camid_dropdown_widget.observe(original_pca_projection.update_data, \"value\")\n",
    "\n",
    "model_pca_projection = ModelPCAProjectionView()\n",
    "camid_dropdown_widget.observe(model_pca_projection.update_data, \"value\")\n",
    "\n",
    "model_edit_pca_projection = ModelEditPCAProjectionView()\n",
    "camid_dropdown_widget.observe(model_edit_pca_projection.update_data, \"value\")\n",
    "edit_btns.observe(model_edit_pca_projection.update_data, \"value\")\n",
    "window_size_dd.observe(model_edit_pca_projection.update_data, \"value\")\n",
    "current_gene_text_widget.observe(model_edit_pca_projection.update_data, \"value\")\n",
    "\n",
    "model_edit_diff_pca_projection = ModelEditDiffPCAProjectionView(\n",
    "    model_pca_projection, model_edit_pca_projection\n",
    ")\n",
    "camid_dropdown_widget.observe(model_edit_diff_pca_projection.update_data, \"value\")\n",
    "edit_btns.observe(model_edit_diff_pca_projection.update_data, \"value\")\n",
    "window_size_dd.observe(model_edit_diff_pca_projection.update_data, \"value\")\n",
    "current_gene_text_widget.observe(model_edit_diff_pca_projection.update_data, \"value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# edit_btns.observe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ipywidgets import TwoByTwoLayout\n",
    "\n",
    "projection_grid = TwoByTwoLayout(\n",
    "    top_left=original_pca_projection.display_box,\n",
    "    top_right=model_pca_projection.display_box,\n",
    "    bottom_left=model_edit_pca_projection.display_box,\n",
    "    bottom_right=model_edit_diff_pca_projection.display_box,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(camid_dropdown_widget)\n",
    "display(manhattan_plot.display_box)\n",
    "display(current_gene_text_widget)\n",
    "display(gene_edit_btns)\n",
    "display(window_size_dd)\n",
    "display(projection_grid)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
