{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading erato-forewings-color_1 Data\n",
      "Length of input data: 484\n",
      "Length of train data: 48\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/carlyn.1/miniconda3/envs/dna/lib/python3.11/site-packages/torch/nn/modules/conv.py:456: UserWarning: Using padding='same' with even kernel lengths and odd dilation may require a zero-padded copy of the input be created (Triggered internally at /opt/conda/conda-bld/pytorch_1699449201450/work/aten/src/ATen/native/Convolution.cpp:1008.)\n",
      "  return F.conv2d(input, weight, bias, self.stride,\n",
      "/home/carlyn.1/miniconda3/envs/dna/lib/python3.11/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.\n",
      "  warnings.warn(\n",
      "/home/carlyn.1/miniconda3/envs/dna/lib/python3.11/site-packages/captum/attr/_core/guided_backprop_deconvnet.py:64: UserWarning: Setting backward hooks on ReLU activations.The hooks will be removed after the attribution is finished\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of input data: 484\n",
      "Length of train data: 48\n",
      "Length of input data: 484\n",
      "Length of train data: 48\n",
      "Length of input data: 484\n",
      "Length of train data: 48\n",
      "Plotting erato-forewings-color_1 Data\n",
      "0.71061206 0.0\n",
      "0.8449885 0.0\n",
      "0.7367694 0.0\n",
      "0.70763004 0.0\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path.append(\"../\")\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "import torch\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.interpolate import splrep, splev, interp1d\n",
    "from tqdm import tqdm\n",
    "\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "from evaluation import get_guided_gradcam_attr, get_saliency_attr, get_guided_gradcam_attr_test\n",
    "from experiments import get_all_large_genes, get_experiment, get_root_paths\n",
    "from data_tools import parse_patternize_csv, load_json, VCF_Dataset\n",
    "from models.net import SoyBeanNet\n",
    "\n",
    "#ROOT = Path(\"/home\",\"carlyn.1\",\"dna-trait-analysis\",\"results\",\"chromosome_out_2\")\n",
    "ROOT = Path(\"/home\",\"carlyn.1\",\"dna-trait-analysis\",\"results\",\"chromosome\")\n",
    "SPECIES = [\"erato\"]\n",
    "#SPECIES = [\"erato\", \"melpomene\"]\n",
    "WINGS = [\"forewings\"]\n",
    "#WINGS = [\"forewings\", \"hindwings\"]\n",
    "COLORS = [\"color_1\", \"color_2\", \"color_3\", \"total\"]\n",
    "#COLORS = [\"color_3\"]\n",
    "DATA_FOLDER = Path(\"melpomene_Hmel218003o_forewings_color_3\")\n",
    "REDO_ATTRIBUTION = True\n",
    "OUT_DIMS = [0]\n",
    "MODEL_OUT_DIMS = 1\n",
    "\n",
    "PLT_COLORS = [\"#767c6f\", \"#373021\", \"#bb2f28\", \"green\"]\n",
    "\n",
    "def load_data(datafile, experiment, batch_size=32):\n",
    "    pca_data = parse_patternize_csv(experiment.pca_loading_path)\n",
    "    input_data = np.load(experiment.gene_vcf_path)['arr_0']\n",
    "    metadata = load_json(experiment.metadata_path)\n",
    "    \n",
    "    cam_ids = []\n",
    "    with open(datafile, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            new_ids = line.split(\",\")\n",
    "            new_ids[-1] = new_ids[-1].strip()\n",
    "            cam_ids.extend(new_ids)\n",
    "    \n",
    "    test_data = []\n",
    "    print(f\"Length of input data: {len(input_data)}\")\n",
    "    for name, row in zip(metadata, input_data):\n",
    "        if name+\"_d\" in pca_data and name in cam_ids:\n",
    "            test_data.append([name, row, pca_data[name+\"_d\"]])\n",
    "    print(f\"Length of train data: {len(test_data)}\")\n",
    "    \n",
    "    test_dataset = VCF_Dataset(test_data)\n",
    "    test_dataloader = DataLoader(test_dataset, batch_size=batch_size, num_workers=8, shuffle=False)\n",
    "    \n",
    "    return test_dataloader, test_data[0][1].shape[0]\n",
    "\n",
    "def load_model(model_path, d_dim, out_dims=2):\n",
    "    model = SoyBeanNet(window_size=d_dim, \n",
    "                       num_out_dims=out_dims, \n",
    "                       insize=3, \n",
    "                       hidden_dim=10, \n",
    "                       drop_out_prob=0.75).cuda()\n",
    "    weights = torch.load(model_path)\n",
    "    model.load_state_dict(weights)\n",
    "    return model\n",
    "\n",
    "def get_chromo_info(path, index):\n",
    "    with open(path, 'r') as f:\n",
    "        line = f.readlines()[index]\n",
    "        line = line.strip()\n",
    "        cols = line.split(\"\\t\")\n",
    "        chromo, pos, ref_alle, alt_alle = cols[:4]\n",
    "        return chromo, pos, ref_alle, alt_alle\n",
    "\n",
    "def get_chromosome_positions(species, gene, wing, color):\n",
    "    vcf_root, _ = get_root_paths()\n",
    "    chrom_path = Path(vcf_root) / Path(f\"{species}/chrom/\" + gene + \".tsv\")\n",
    "    with open(chrom_path, 'r') as f:\n",
    "        positions = [int(line.strip().split(\"\\t\")[1]) for line in f.readlines()]\n",
    "\n",
    "    return positions\n",
    "\n",
    "def format_title(species, wing, color):\n",
    "    color_txt = ' '.join(color.capitalize().split(\"_\"))\n",
    "    return f\"{species.capitalize()} {wing.capitalize()} {color_txt}\"\n",
    "    \n",
    "large_genes = get_all_large_genes()\n",
    "for species in SPECIES:\n",
    "    for wing in WINGS:\n",
    "        for c, color in enumerate(COLORS):\n",
    "            scaled_data = []\n",
    "            rmse_data = None\n",
    "            print(f\"Loading {species}-{wing}-{color} Data\")\n",
    "            for i, gene in enumerate(large_genes[species]):\n",
    "                exp_name = \"_\".join([species, gene, wing, color])\n",
    "                data_folder = ROOT / Path(exp_name)\n",
    "                with open(data_folder / \"out.log\", 'r') as f:\n",
    "                    lines = f.readlines()\n",
    "                    parts = lines[-3].strip().split(\" \")\n",
    "                    train_rmse = float(parts[2])\n",
    "                    val_rmse = float(parts[6])\n",
    "                    test_rmse = float(parts[10])\n",
    "                    rmse = np.array([[train_rmse, val_rmse, test_rmse]])\n",
    "                    if i == 0:\n",
    "                        rmse_data = rmse\n",
    "                    else:\n",
    "                        rmse_data = np.concatenate((rmse_data, rmse), axis=0)\n",
    "                    \n",
    "                if not REDO_ATTRIBUTION:\n",
    "                    data_path = data_folder / \"att_points.npz\"\n",
    "                    test_att = np.load(data_path)[\"test\"]\n",
    "                else:\n",
    "                    model_path = data_folder / \"model.pt\"\n",
    "                    test_datafile = data_folder / \"test_split.txt\"\n",
    "                    experiment = get_experiment(species, gene, wing, color, is_large=True)\n",
    "                    dloader, d_dim = load_data(test_datafile, experiment, batch_size=4)\n",
    "                    torch.cuda.set_device(2)\n",
    "                    model = load_model(model_path, d_dim, out_dims=MODEL_OUT_DIMS)\n",
    "                    model.eval()\n",
    "                    test_att = None\n",
    "                    for out_tgt in OUT_DIMS:\n",
    "                        if False:\n",
    "                            att = get_saliency_attr(model, dloader, target=out_tgt)\n",
    "                        else:\n",
    "                            #att = get_guided_gradcam_attr(model, dloader, target=out_tgt)\n",
    "                            att = get_guided_gradcam_attr_test(model, dloader, target=out_tgt)\n",
    "                        if test_att is None:\n",
    "                            test_att = att\n",
    "                        else:\n",
    "                            test_att += att\n",
    "                        \n",
    "                    \n",
    "                test_att = np.abs(test_att)\n",
    "                test_att -= test_att.min()\n",
    "                test_att /= test_att.max()\n",
    "                scaled_data.append(test_att)\n",
    "\n",
    "            \n",
    "            # Scale the data\n",
    "            idx = 1 # Scaling based on the validation RMSE\n",
    "            gene_scale = rmse_data[:, idx] / rmse_data[:, idx].sum() # We normalize this way, as the highest RMSE should not get zeros in the end\n",
    "            gene_scale = 1 - gene_scale # We flip as we want lower RMSE to have higher contribution\n",
    "            #gene_scale = [1]\n",
    "            #assert len(gene_scale) == 4, \"There should only be 4 for the 4 genes.\"            \n",
    "\n",
    "            print(f\"Plotting {species}-{wing}-{color} Data\")\n",
    "            fig, axs = plt.subplots(4, 1, figsize=(20, 40), dpi=300)\n",
    "            #fig_title = f\"{species} - {wing} - {color}\"\n",
    "            fig_title = format_title(species, wing, color)\n",
    "            fig.suptitle(fig_title, fontsize=40)\n",
    "            FONT_SIZE = 18\n",
    "            plt.rc('font', size=FONT_SIZE)            # fontsize of the text sizes\n",
    "            plt.rc('legend', fontsize=FONT_SIZE)  # fontsize of the legend\n",
    "            plt.rc('axes', labelsize='medium', titlesize='large')\n",
    "            for i, gene in enumerate(large_genes[species]):\n",
    "                chromosome_positions = get_chromosome_positions(species, gene, wing, color)\n",
    "                \n",
    "                TMP = True\n",
    "                if TMP:\n",
    "                    fig, ax = plt.subplots(1, 1, figsize=(20, 10), dpi=300)\n",
    "                    #fig_title = f\"{species} - {wing} - {color}\"\n",
    "                    fig_title = format_title(species, wing, color)\n",
    "                    fig.suptitle(fig_title, fontsize=40)\n",
    "                else:\n",
    "                    ax = axs[i]\n",
    "                \n",
    "                test_att = scaled_data[i]\n",
    "                # Scale\n",
    "                test_att *= gene_scale[i]\n",
    "                \n",
    "                ax.set_ylabel('Attribution', fontsize=20)\n",
    "                ax.set_xlabel('SNP Position', fontsize=20)\n",
    "                ax.tick_params(axis='both', labelsize=14)\n",
    "                #ax.set_title(f\"Chromosome: {gene} - RMSE: {round(rmse_data[i, idx], 4)}\")\n",
    "                ax.set_title(f\"Chromosome: {gene}\", fontsize=28)\n",
    "\n",
    "                x_type = \"chromosome\"\n",
    "                if x_type == \"chromosome\":\n",
    "                    #x_max = max(chromosome_positions)\n",
    "                    X = np.array(chromosome_positions)\n",
    "                else:\n",
    "                    x_max = test_att.shape[0]\n",
    "                    X = np.arange(x_max)\n",
    "\n",
    "                y = test_att\n",
    "                ys = (y - y.min())/(y.max()-y.min())\n",
    "                alpha = 0.2 + ys * 0.8\n",
    "                print(y.max(), y.min())\n",
    "                s = ys * 20\n",
    "\n",
    "                ax.scatter(X, y, alpha=alpha, s=s, color=PLT_COLORS[c])\n",
    "                SMOOTH = False\n",
    "                SMOOTH_ALPHA = 0.50\n",
    "                if SMOOTH:\n",
    "                    smooth_y = []\n",
    "                    for i, (orgx, orgy) in enumerate(zip(X, y)):\n",
    "                        if i == 0:\n",
    "                            smooth_y.append(orgy)\n",
    "                            continue\n",
    "                        \n",
    "                        newy = SMOOTH_ALPHA * orgy + (1-SMOOTH_ALPHA) * smooth_y[-1]\n",
    "                        smooth_y.append(newy)\n",
    "                        \n",
    "                    #spline_fn = splrep(chromosome_positions, y)\n",
    "                    #spline_fn = interp1d(chromosome_positions, y, kind=\"cubic\")\n",
    "                    #spline_X = np.linspace(min(chromosome_positions), max(chromosome_positions), 50)\n",
    "                    #spline_Y = spline_fn(spline_X)\n",
    "                    #spline_Y = splev(spline_X, spline_fn)\n",
    "                    #spline_Y = spline_fn(spline_X)\n",
    "                    #ax.plot(spline_X, spline_Y, color=\"brown\", alpha=0.5, linewidth=3)\n",
    "                    ax.plot(X, smooth_y, color=\"brown\", alpha=0.5, linewidth=3)\n",
    "                #ax.bar(X[prev:prev+gd.size], test_att[prev:prev+gd.size], align='center', alpha=0.1, s=0.1, color=colors[i], label=gd.gene)\n",
    "                #prev += gd.size\n",
    "                #handles.append(mpatches.Patch(color=colors[i], label=gd.gene))\n",
    "\n",
    "                #ax.set_ylim([0, 1.1])\n",
    "                #ax.legend(handles=handles)\n",
    "                # On the higher end, takes about 16s\n",
    "                \n",
    "                # Get max position\n",
    "                max_positions = np.argsort(y)[-5:][::-1]\n",
    "                #max_positions = []\n",
    "                prev_plots = []\n",
    "                \n",
    "                def too_close(new_pos):\n",
    "                    for p in prev_plots:\n",
    "                        diff = abs(new_pos - p)\n",
    "                        if  diff < 10_000: return True\n",
    "                    return False\n",
    "                \n",
    "                for max_i in max_positions:\n",
    "                    if too_close(max_i): continue # Don't plot too close\n",
    "                    prev_plots.append(max_i)\n",
    "                    chromo_pos = chromosome_positions[max_i]\n",
    "                    ax.annotate(chromo_pos,\n",
    "                                xy=(X[max_i], y[max_i]),\n",
    "                                xycoords=\"data\",\n",
    "                                xytext=(X[max_i], y[max_i]),\n",
    "                                arrowprops=dict(facecolor=PLT_COLORS[c], shrink=0.05),\n",
    "                                horizontalalignment='center',\n",
    "                                verticalalignment='bottom')\n",
    "\n",
    "                with open(f\"chrom_{gene}_{fig_title}.csv\", 'w') as f:\n",
    "                    f.write(\"Chromosome Position,Attribution\\n\")\n",
    "                    for pos, att in zip(chromosome_positions, y):\n",
    "                        f.write(f\"{pos},{att}\\n\")\n",
    "                        \n",
    "                if TMP:\n",
    "                    fig.tight_layout()\n",
    "                    fig.subplots_adjust(top=0.85) # Adjusts the figure title so it isn't cramped into the subplot titles\n",
    "                    plt.savefig(f\"chrom_{fig_title}_{gene}.png\")\n",
    "            fig.tight_layout()\n",
    "            fig.subplots_adjust(top=0.95) # Adjusts the figure title so it isn't cramped into the subplot titles\n",
    "            plt.savefig(f\"chrom_{fig_title}.png\")\n",
    "            \n",
    "            \n",
    "# To plot all figures, it takes: 10 - 15min\n",
    "# To plot one figure, for one row: 48 seconds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dna",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
