{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "gwas_points_file_path = Path(\"/local/scratch/carlyn.1/dna/metadata/gwas_Herato1001_4570000_4850000.txt\")\n",
    "gwas_df = pd.read_csv(gwas_points_file_path, delimiter=\"\\t\")\n",
    "\n",
    "annotation_file_path = Path(\"/local/scratch/carlyn.1/dna/metadata/wnta_annotation_mel2x5_eratodem1.xlsx\")\n",
    "anno_df = pd.read_excel(annotation_file_path,  sheet_name=\"Complete_wnta_annt_demopho\")\n",
    "anno_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_DIR = Path(\"/local/scratch/carlyn.1/dna/training_output/pca_10/\")\n",
    "DNA_ROOT_DIR = Path(\"/local/scratch/carlyn.1/dna/processed/genome/\")\n",
    "EDIT_OPTIONS = [\"AA\", \"Aa/aA\", \"aa\", \"zero-out\"]\n",
    "PHASE_NAMES = [\"training\", \"validation\", \"test\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from gtp.dataloading.tools import load_json, collect_chromosome_position_metadata\n",
    "\n",
    "\n",
    "def load_data(exp_dir, species, wing, color, chromosome):\n",
    "    position_metadata = collect_chromosome_position_metadata(\n",
    "        DNA_ROOT_DIR, species, chromosome\n",
    "    )\n",
    "    data_dir = exp_dir / f\"{species}_{wing}_{color}_chromosome_{chromosome}\"\n",
    "    all_data = []\n",
    "    for phase in [\"training\", \"validation\", \"test\"]:\n",
    "        data = np.load(\n",
    "            data_dir / f\"{phase}_windowed_editing_attributions.npy\", allow_pickle=True\n",
    "        ).item()\n",
    "        metrics = load_json(data_dir / f\"{phase}_metrics.json\")\n",
    "        pearsonr = metrics[\"pearsonr\"]\n",
    "        rmse = metrics[\"rmse\"]\n",
    "        all_data.append([data, rmse, pearsonr])\n",
    "\n",
    "    return all_data, position_metadata\n",
    "\n",
    "\n",
    "def collect_plot_data(data, nominal_positions):\n",
    "    plot_data = {}\n",
    "    for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "        all_pos = []\n",
    "        for pos in nominal_positions:\n",
    "            all_data = np.concatenate(data[pos][i])\n",
    "            all_pos.append(all_data)\n",
    "        plot_data[edit_opt] = np.stack(all_pos, axis=1)\n",
    "    return plot_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_pos_min = anno_df.Start.min()\n",
    "bp_pos_max = anno_df.End.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_df.Type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "\n",
    "species = \"erato\"\n",
    "wing = \"forewings\"\n",
    "color = \"color_3\"\n",
    "chromosome = 10\n",
    "all_phase_data, position_metadata = load_data(EXP_DIR, species=species, wing=wing, color=color, chromosome=10)\n",
    "\n",
    "all_nominal_positions = [list(all_phase_data[i][0].keys()) for i in range(3)]\n",
    "#assert (\n",
    "#    all_nominal_positions[0] == all_nominal_positions[1] == all_nominal_positions[2]\n",
    "#), \"All position must match across phases.\"\n",
    "\n",
    "chromosome_position_metadata = []\n",
    "chromosome_position_metadata_start_end = []\n",
    "for nom_pos in all_nominal_positions[2]:\n",
    "    if nom_pos == len(position_metadata):\n",
    "        nom_pos -= 1\n",
    "    chromosome_position_metadata.append(position_metadata[nom_pos])\n",
    "    chromosome_position_metadata_start_end.append([position_metadata[max(0, nom_pos-10)], position_metadata[min(len(position_metadata)-1, nom_pos+10)]])\n",
    "\n",
    "phase_data = all_phase_data[PHASE_NAMES.index(\"test\")]\n",
    "attrs, rmse, pearsonr = phase_data\n",
    "nominal_window_positions = list(attrs.keys())\n",
    "plot_data = collect_plot_data(attrs, nominal_window_positions)\n",
    "\n",
    "all_real_positions = [x[1] for x in chromosome_position_metadata]\n",
    "all_real_positions_start_and_end = [[x1[1], x2[1]] for x1, x2 in chromosome_position_metadata_start_end]\n",
    "\n",
    "start_pos = None\n",
    "end_pos = all_real_positions[-1]\n",
    "for i, rp in enumerate(all_real_positions):\n",
    "    if start_pos is None and rp > bp_pos_min:\n",
    "        start_pos = max(0, i-1)\n",
    "    if rp > bp_pos_max:\n",
    "        end_pos = i\n",
    "        break\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_data[\"AA\"].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = 0.0005\n",
    "max_height = 0\n",
    "fig, ax = plt.subplots(1, 1, figsize=(24, 9))\n",
    "for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "    means = np.abs(plot_data[edit_opt]).mean(axis=0)\n",
    "    y = means[start_pos:end_pos+1]\n",
    "    max_height = max(max(y), max_height)\n",
    "    for j, (s, e) in enumerate(all_real_positions_start_and_end[start_pos:end_pos+1]):\n",
    "        v = y[j]\n",
    "        rect = patches.Rectangle((s, v-(w/2)), e-s, w, linewidth=1, edgecolor='black', facecolor='purple', alpha=0.1)\n",
    "        ax.add_patch(rect)\n",
    "    ax.scatter(all_real_positions[start_pos:end_pos+1], means[start_pos:end_pos+1], label=edit_opt, s=10)\n",
    "    \n",
    "for i, row in anno_df.iterrows():\n",
    "    y = max_height + w * 10\n",
    "    if row[\"Type\"] == \"gene\":\n",
    "        y -= w*2\n",
    "        c = \"blue\"\n",
    "    elif row[\"Type\"] == \"phenotype\":\n",
    "        y -= w*4\n",
    "        c = \"yellow\"\n",
    "    elif row[\"Type\"] == \"atacSeq peak\":\n",
    "        y -= w*6\n",
    "        c = \"red\"\n",
    "    elif row[\"Type\"] == \"exon\":\n",
    "        c = \"green\"\n",
    "        y -= w*8\n",
    "    rect = patches.Rectangle((row.Start, y), row.End-row.Start, w, linewidth=1, edgecolor='black', facecolor=c)\n",
    "    ax.add_patch(rect)\n",
    "\n",
    "ax.legend()\n",
    "    \n",
    "plt.grid()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "x = gwas_df[\"ps\"]\n",
    "y = gwas_df[\"neg_log10_p\"]\n",
    "y -= y.min()\n",
    "y /= y.max()\n",
    "y *= max_height\n",
    "\n",
    "gobj = go.Scatter(\n",
    "    x=x, \n",
    "    y=y, \n",
    "    name=\"GWAS\", \n",
    "    mode='markers', \n",
    "    marker=dict(size=5, color='black'),\n",
    ")\n",
    "fig.add_trace(gobj)\n",
    "\n",
    "\n",
    "for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "    means = np.abs(plot_data[edit_opt]).mean(axis=0)\n",
    "    y = means[start_pos:end_pos+1]\n",
    "    x = all_real_positions[start_pos:end_pos+1]\n",
    "    s, e = list(zip(*all_real_positions_start_and_end[start_pos:end_pos+1]))\n",
    "    error_x = np.array(e) - np.array(x)\n",
    "    error_x_minus = np.array(x) - np.array(s)\n",
    "    gobj = go.Scatter(\n",
    "        x=x, \n",
    "        y=y, \n",
    "        name=edit_opt, \n",
    "        mode='markers', \n",
    "        marker=dict(size=5), \n",
    "        error_x=dict(type=\"data\", array=error_x, arrayminus=error_x_minus),\n",
    "    )\n",
    "    fig.add_trace(gobj)\n",
    "\n",
    "for i, row in anno_df.iterrows():\n",
    "    y = max_height + w * 10\n",
    "    if row[\"Type\"] == \"gene\":\n",
    "        y -= w*2\n",
    "        c = \"blue\"\n",
    "    elif row[\"Type\"] == \"phenotype\":\n",
    "        y -= w*4\n",
    "        c = \"yellow\"\n",
    "    elif row[\"Type\"] == \"atacSeq peak\":\n",
    "        y -= w*6\n",
    "        c = \"red\"\n",
    "    elif row[\"Type\"] == \"exon\":\n",
    "        c = \"green\"\n",
    "        y -= w*8\n",
    "    fig.add_shape(\n",
    "        type=\"rect\",\n",
    "        xref=\"x\",  # Reference x-axis data coordinates\n",
    "        yref=\"y\",  # Reference y-axis data coordinates\n",
    "        x0=row.Start,    # Starting x-coordinate of the rectangle\n",
    "        y0=y-(w/2),    # Starting y-coordinate of the rectangle\n",
    "        x1=row.End,    # Ending x-coordinate of the rectangle\n",
    "        y1=y+(w/2),    # Ending y-coordinate of the rectangle\n",
    "        line=dict(\n",
    "            color=c,\n",
    "            width=0.002,\n",
    "        ),\n",
    "        fillcolor=c,\n",
    "        opacity=1,\n",
    "        layer=\"below\" # Place the rectangle below the data points\n",
    "    )\n",
    "    \n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=[row.Start + (row.End-row.Start)/2], \n",
    "            y=[y],\n",
    "            name=row.Type,\n",
    "            hovertemplate=f\"{row.Type} {row.name} : {row.Start}-{row.End}\",\n",
    "            showlegend=False,\n",
    "            mode='markers', \n",
    "            marker=dict(color=\"black\", size=1),\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig.update_layout(dict(hovermode='closest', xaxis=dict(showspikes=True), xaxis_spikemode='across'))\n",
    "\n",
    "fig.write_html(f\"../output/results/{species}_{wing}_{color}_{chromosome}.html\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_dfs = []\n",
    "for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "    means = np.abs(plot_data[edit_opt]).mean(axis=0)\n",
    "    y = means[start_pos:end_pos+1]\n",
    "    x = all_real_positions[start_pos:end_pos+1]\n",
    "    \n",
    "    df = pd.DataFrame(np.stack((x, y), axis=1), columns=[\"Position\", \"Attribution\"])\n",
    "    df[\"Edit_Type\"] = edit_opt\n",
    "    df.Position = df.Position.astype(int)\n",
    "    all_dfs.append(df)\n",
    "\n",
    "all_df = pd.concat(all_dfs)\n",
    "\n",
    "all_df.to_csv(f\"../output/results/{species}_{wing}_{color}_{chromosome}.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
