{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading erato-forewings-color_3 Data\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/carlyn.1/dna-trait-analysis/results/dim_2_wntA_total/erato_Herato1001_wntA_forewings_color_3/out.log'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/carlyn.1/dna-trait-analysis/notebooks/plot_large_chromosome_single.ipynb Cell 1\u001b[0m line \u001b[0;36m9\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_large_chromosome_single.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=96'>97</a>\u001b[0m exp_name \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m_\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin([SPECIES, GENE, WING, COLOR])\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_large_chromosome_single.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=97'>98</a>\u001b[0m data_folder \u001b[39m=\u001b[39m ROOT \u001b[39m/\u001b[39m Path(exp_name)\n\u001b[0;32m---> <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_large_chromosome_single.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=98'>99</a>\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39;49m(data_folder \u001b[39m/\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39mout.log\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mr\u001b[39;49m\u001b[39m'\u001b[39;49m) \u001b[39mas\u001b[39;00m f:\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_large_chromosome_single.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=99'>100</a>\u001b[0m     lines \u001b[39m=\u001b[39m f\u001b[39m.\u001b[39mreadlines()\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_large_chromosome_single.ipynb#W0sdnNjb2RlLXJlbW90ZQ%3D%3D?line=100'>101</a>\u001b[0m     parts \u001b[39m=\u001b[39m lines[\u001b[39m-\u001b[39m\u001b[39m3\u001b[39m]\u001b[39m.\u001b[39mstrip()\u001b[39m.\u001b[39msplit(\u001b[39m\"\u001b[39m\u001b[39m \u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/dna/lib/python3.12/site-packages/IPython/core/interactiveshell.py:324\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    317\u001b[0m \u001b[39mif\u001b[39;00m file \u001b[39min\u001b[39;00m {\u001b[39m0\u001b[39m, \u001b[39m1\u001b[39m, \u001b[39m2\u001b[39m}:\n\u001b[1;32m    318\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    319\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mIPython won\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt let you open fd=\u001b[39m\u001b[39m{\u001b[39;00mfile\u001b[39m}\u001b[39;00m\u001b[39m by default \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    320\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    321\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39myou can use builtins\u001b[39m\u001b[39m'\u001b[39m\u001b[39m open.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    322\u001b[0m     )\n\u001b[0;32m--> 324\u001b[0m \u001b[39mreturn\u001b[39;00m io_open(file, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/carlyn.1/dna-trait-analysis/results/dim_2_wntA_total/erato_Herato1001_wntA_forewings_color_3/out.log'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path.append(\"../\")\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "import torch\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.interpolate import splrep, splev, interp1d\n",
    "from tqdm import tqdm\n",
    "\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "from gtp.evaluation import get_guided_gradcam_attr, get_saliency_attr, get_guided_gradcam_attr_test\n",
    "from gtp.experiments import get_all_large_genes, get_experiment, get_root_paths\n",
    "from gtp.data_tools import parse_patternize_csv, load_json, VCF_Dataset\n",
    "from gtp.models.net import SoyBeanNet\n",
    "\n",
    "RESULTS_DIR = Path(\"/home/carlyn.1/dna-trait-analysis/plot_results\")\n",
    "\n",
    "#ROOT = Path(\"/home\",\"carlyn.1\",\"dna-trait-analysis\",\"results\",\"chromosome_out_2\")\n",
    "ROOT = Path(\"/home\",\"carlyn.1\",\"dna-trait-analysis\",\"results\",\"dim_2_wntA_total\")\n",
    "SPECIES = \"erato\"\n",
    "#SPECIES = [\"erato\", \"melpomene\"]\n",
    "WING = \"forewings\"\n",
    "#WINGS = [\"forewings\", \"hindwings\"]\n",
    "COLOR = \"total\"\n",
    "GENE = \"Herato1001_wntA\"\n",
    "#COLORS = [\"color_3\"]\n",
    "#DATA_FOLDER = Path(\"melpomene_Hmel218003o_forewings_color_3\")\n",
    "REDO_ATTRIBUTION = True\n",
    "OUT_DIMS = [0]\n",
    "MODEL_OUT_DIMS = 1\n",
    "\n",
    "PLT_COLORS = [\"#767c6f\", \"#373021\", \"#bb2f28\", \"green\"]\n",
    "\n",
    "def load_data(datafile, experiment, batch_size=32):\n",
    "    pca_data = parse_patternize_csv(experiment.pca_loading_path)\n",
    "    input_data = np.load(experiment.gene_vcf_path)['arr_0']\n",
    "    metadata = load_json(experiment.metadata_path)\n",
    "    \n",
    "    cam_ids = []\n",
    "    with open(datafile, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            new_ids = line.split(\",\")\n",
    "            new_ids[-1] = new_ids[-1].strip()\n",
    "            cam_ids.extend(new_ids)\n",
    "    \n",
    "    test_data = []\n",
    "    print(f\"Length of input data: {len(input_data)}\")\n",
    "    for name, row in zip(metadata, input_data):\n",
    "        if name+\"_d\" in pca_data and name in cam_ids:\n",
    "            test_data.append([name, row, pca_data[name+\"_d\"]])\n",
    "    print(f\"Length of train data: {len(test_data)}\")\n",
    "    \n",
    "    test_dataset = VCF_Dataset(test_data)\n",
    "    test_dataloader = DataLoader(test_dataset, batch_size=batch_size, num_workers=8, shuffle=False)\n",
    "    \n",
    "    return test_dataloader, test_data[0][1].shape[0]\n",
    "\n",
    "def load_model(model_path, d_dim, out_dims=2):\n",
    "    model = SoyBeanNet(window_size=d_dim, \n",
    "                       num_out_dims=out_dims, \n",
    "                       insize=3, \n",
    "                       hidden_dim=10, \n",
    "                       drop_out_prob=0.75).cuda()\n",
    "    weights = torch.load(model_path)\n",
    "    model.load_state_dict(weights)\n",
    "    return model\n",
    "\n",
    "def get_chromo_info(path, index):\n",
    "    with open(path, 'r') as f:\n",
    "        line = f.readlines()[index]\n",
    "        line = line.strip()\n",
    "        cols = line.split(\"\\t\")\n",
    "        chromo, pos, ref_alle, alt_alle = cols[:4]\n",
    "        return chromo, pos, ref_alle, alt_alle\n",
    "\n",
    "def get_chromosome_positions(species, gene, wing, color):\n",
    "    vcf_root, _ = get_root_paths()\n",
    "    chrom_path = Path(vcf_root) / Path(f\"{species}/\" + gene + \".tsv\")\n",
    "    with open(chrom_path, 'r') as f:\n",
    "        positions = [int(line.strip().split(\"\\t\")[1]) for line in f.readlines()]\n",
    "\n",
    "    return positions\n",
    "\n",
    "def format_title(species, wing, color):\n",
    "    color_txt = ' '.join(color.capitalize().split(\"_\"))\n",
    "    return f\"{species.capitalize()} {wing.capitalize()} {color_txt}\"\n",
    "    \n",
    "\n",
    "scaled_data = []\n",
    "rmse_data = None\n",
    "print(f\"Loading {SPECIES}-{WING}-{COLOR} Data\")\n",
    "exp_name = \"_\".join([SPECIES, GENE, WING, COLOR])\n",
    "data_folder = ROOT / Path(exp_name)\n",
    "with open(data_folder / \"out.log\", 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    parts = lines[-3].strip().split(\" \")\n",
    "    #train_rmse = float(parts[2])\n",
    "    #val_rmse = float(parts[6])\n",
    "    #test_rmse = float(parts[10])\n",
    "    train_rmse = -1\n",
    "    val_rmse = -1\n",
    "    test_rmse = -1\n",
    "    \n",
    "    rmse = np.array([[train_rmse, val_rmse, test_rmse]])\n",
    "    rmse_data = rmse\n",
    "\n",
    "    \n",
    "if not REDO_ATTRIBUTION:\n",
    "    data_path = data_folder / \"att_points.npz\"\n",
    "    test_att = np.load(data_path)[\"test\"]\n",
    "else:\n",
    "    model_path = data_folder / \"model.pt\"\n",
    "    test_datafile = data_folder / \"test_split.txt\"\n",
    "    experiment = get_experiment(SPECIES, GENE, WING, COLOR, is_large=False)\n",
    "    dloader, d_dim = load_data(test_datafile, experiment, batch_size=4)\n",
    "    torch.cuda.set_device(2)\n",
    "    model = load_model(model_path, d_dim, out_dims=MODEL_OUT_DIMS)\n",
    "    model.eval()\n",
    "    test_att = None\n",
    "    for out_tgt in OUT_DIMS:\n",
    "        if True:\n",
    "            att = get_saliency_attr(model, dloader, target=out_tgt)\n",
    "        else:\n",
    "            #att = get_guided_gradcam_attr(model, dloader, target=out_tgt)\n",
    "            att = get_guided_gradcam_attr_test(model, dloader, target=out_tgt)\n",
    "        if test_att is None:\n",
    "            test_att = att\n",
    "        else:\n",
    "            test_att += att\n",
    "        \n",
    "    \n",
    "test_att = np.abs(test_att)\n",
    "test_att -= test_att.min()\n",
    "test_att /= test_att.max()\n",
    "scaled_data = test_att\n",
    "\n",
    "        \n",
    "# Scale the data\n",
    "idx = 1 # Scaling based on the validation RMSE\n",
    "gene_scale = rmse_data[:, idx]\n",
    "#gene_scale = rmse_data[:, idx] / rmse_data[:, idx].sum() # We normalize this way, as the highest RMSE should not get zeros in the end\n",
    "#gene_scale = 1 - gene_scale # We flip as we want lower RMSE to have higher contribution\n",
    "#gene_scale = [1]\n",
    "#assert len(gene_scale) == 4, \"There should only be 4 for the 4 genes.\"            \n",
    "\n",
    "print(f\"Plotting {SPECIES}-{WING}-{COLOR} Data\")\n",
    "#fig, axs = plt.subplots(1, 1, figsize=(15, 10), dpi=300)\n",
    "#fig_title = f\"{SPECIES} - {WING} - {COLOR}\"\n",
    "fig_title = format_title(SPECIES, WING, COLOR)\n",
    "#fig.suptitle(fig_title, fontsize=40)\n",
    "FONT_SIZE = 14\n",
    "plt.rc('font', size=FONT_SIZE)            # fontsize of the text sizes\n",
    "plt.rc('legend', fontsize=FONT_SIZE)  # fontsize of the legend\n",
    "plt.rc('axes', labelsize='medium', titlesize='large')\n",
    "chromosome_positions = get_chromosome_positions(SPECIES, GENE, WING, COLOR)\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(15, 10), dpi=300)\n",
    "#fig_title = f\"{species} - {wing} - {color}\"\n",
    "fig_title = format_title(SPECIES, WING, COLOR)\n",
    "fig.suptitle(fig_title, fontsize=22)\n",
    "\n",
    "test_att = scaled_data\n",
    "# Scale\n",
    "#test_att *= gene_scale[i]\n",
    "\n",
    "ax.set_ylabel('Attribution', fontsize=14)\n",
    "ax.set_xlabel('SNP Position', fontsize=14)\n",
    "ax.tick_params(axis='both', labelsize=10)\n",
    "#ax.set_title(f\"Chromosome: {gene} - RMSE: {round(rmse_data[i, idx], 4)}\")\n",
    "ax.set_title(f\"Chromosome: {GENE}\", fontsize=18)\n",
    "\n",
    "x_type = \"chromosome\"\n",
    "if x_type == \"chromosome\":\n",
    "    #x_max = max(chromosome_positions)\n",
    "    X = np.array(chromosome_positions)\n",
    "else:\n",
    "    x_max = test_att.shape[0]\n",
    "    X = np.arange(x_max)\n",
    "\n",
    "y = test_att\n",
    "ys = (y - y.min())/(y.max()-y.min())\n",
    "alpha = 0.2 + ys * 0.8\n",
    "print(y.max(), y.min())\n",
    "s = ys * 20\n",
    "\n",
    "ax.scatter(X, y, alpha=alpha, s=s, color=PLT_COLORS[2])\n",
    "SMOOTH = False\n",
    "SMOOTH_ALPHA = 0.50\n",
    "if SMOOTH:\n",
    "    smooth_y = []\n",
    "    for i, (orgx, orgy) in enumerate(zip(X, y)):\n",
    "        if i == 0:\n",
    "            smooth_y.append(orgy)\n",
    "            continue\n",
    "        \n",
    "        newy = SMOOTH_ALPHA * orgy + (1-SMOOTH_ALPHA) * smooth_y[-1]\n",
    "        smooth_y.append(newy)\n",
    "        \n",
    "    #spline_fn = splrep(chromosome_positions, y)\n",
    "    #spline_fn = interp1d(chromosome_positions, y, kind=\"cubic\")\n",
    "    #spline_X = np.linspace(min(chromosome_positions), max(chromosome_positions), 50)\n",
    "    #spline_Y = spline_fn(spline_X)\n",
    "    #spline_Y = splev(spline_X, spline_fn)\n",
    "    #spline_Y = spline_fn(spline_X)\n",
    "    #ax.plot(spline_X, spline_Y, color=\"brown\", alpha=0.5, linewidth=3)\n",
    "    ax.plot(X, smooth_y, color=\"brown\", alpha=0.5, linewidth=3)\n",
    "    #ax.bar(X[prev:prev+gd.size], test_att[prev:prev+gd.size], align='center', alpha=0.1, s=0.1, color=colors[i], label=gd.gene)\n",
    "    #prev += gd.size\n",
    "    #handles.append(mpatches.Patch(color=colors[i], label=gd.gene))\n",
    "\n",
    "    #ax.set_ylim([0, 1.1])\n",
    "    #ax.legend(handles=handles)\n",
    "    # On the higher end, takes about 16s\n",
    "    \n",
    "    # Get max position\n",
    "    max_positions = np.argsort(y)[-5:][::-1]\n",
    "    #max_positions = []\n",
    "    prev_plots = []\n",
    "    \n",
    "    def too_close(new_pos):\n",
    "        for p in prev_plots:\n",
    "            diff = abs(new_pos - p)\n",
    "            if  diff < 10_000: return True\n",
    "        return False\n",
    "    \n",
    "    for max_i in max_positions:\n",
    "        if too_close(max_i): continue # Don't plot too close\n",
    "        prev_plots.append(max_i)\n",
    "        chromo_pos = chromosome_positions[max_i]\n",
    "        ax.annotate(chromo_pos,\n",
    "                    xy=(X[max_i], y[max_i]),\n",
    "                    xycoords=\"data\",\n",
    "                    xytext=(X[max_i], y[max_i]),\n",
    "                    arrowprops=dict(facecolor=PLT_COLORS[2], shrink=0.05),\n",
    "                    horizontalalignment='center',\n",
    "                    verticalalignment='bottom')\n",
    "\n",
    "    with open(f\"chrom_{GENE}_{fig_title}.csv\", 'w') as f:\n",
    "        f.write(\"Chromosome Position,Attribution\\n\")\n",
    "        for pos, att in zip(chromosome_positions, y):\n",
    "            f.write(f\"{pos},{att}\\n\")\n",
    "            \n",
    "    fig.tight_layout()\n",
    "    fig.subplots_adjust(top=0.85) # Adjusts the figure title so it isn't cramped into the subplot titles\n",
    "    plt.savefig(RESULTS_DIR / f\"chrom_{fig_title}_{GENE}.png\")\n",
    "    \n",
    "else:\n",
    "    fig.tight_layout()\n",
    "    fig.subplots_adjust(top=0.85) # Adjusts the figure title so it isn't cramped into the subplot titles\n",
    "    plt.savefig(RESULTS_DIR / f\"chrom_{fig_title}.png\")\n",
    "            \n",
    "            \n",
    "# To plot all figures, it takes: 10 - 15min\n",
    "# To plot one figure, for one row: 48 seconds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dna",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
