{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "PHENOTYPE_FOLDER = Path(\"/local/scratch/carlyn.1/dna/colors/processed\")\n",
    "GENOME_FOLDER = Path(\"/local/scratch/carlyn.1/dna/vcfs/processed/genome\")\n",
    "DATA_SPLIT_FOLDER = Path(\"/home/carlyn.1/dna-trait-analysis/data\")\n",
    "RESULTS_FOLDER = Path(\"/local/scratch/carlyn.1/dna/results/whole_genome\")\n",
    "PLOT_RESULTS_FOLDER = Path(\"/local/scratch/carlyn.1/dna/results/plot_whole_genome\")\n",
    "BATCH_SIZE = 32\n",
    "DEVICE = \"cuda:1\"\n",
    "REDO = False\n",
    "\n",
    "ex_path = \"/local/scratch/carlyn.1/dna/results/whole_genome/erato_forewings_total_chromosome_10\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "from captum.attr import GuidedGradCam\n",
    "\n",
    "def collect_chromosome(root, species, chromosome):\n",
    "    genotype_path_root = os.path.join(root, f\"{species}\")\n",
    "    scaffolds = []\n",
    "    for subdir in os.listdir(genotype_path_root):\n",
    "        if species == \"erato\":\n",
    "            chrom_info = subdir.replace(\"Herato\", \"\")\n",
    "            chrom_num = int(chrom_info[:2])\n",
    "            scaf_num = int(chrom_info[2:])\n",
    "        elif species == \"melpomene\":\n",
    "            chrom_info = subdir.replace(\"Hmel2\", \"\")\n",
    "            chrom_num = int(chrom_info[:2])\n",
    "            scaf_num = int(chrom_info[2:-1])\n",
    "        else:\n",
    "            assert False, f\"Invalid species: {species}\"\n",
    "            \n",
    "        if chrom_num == chromosome:\n",
    "            scaffolds.append([subdir, scaf_num])\n",
    "    scaffolds = sorted(scaffolds, key=lambda x: x[1])\n",
    "\n",
    "    sort_verify = None\n",
    "    compiled = None\n",
    "    for scaffold_dir, _ in scaffolds:\n",
    "        tgt_dir = os.path.join(genotype_path_root, scaffold_dir)\n",
    "        camids = np.load(os.path.join(tgt_dir, \"camids.npy\"))\n",
    "        data = np.load(os.path.join(tgt_dir, \"ml_ready.npy\"))\n",
    "        sort_idx = np.argsort(camids)\n",
    "        sorted_data = data[sort_idx].astype(np.byte)\n",
    "        sorted_camids = camids[sort_idx]\n",
    "        if sort_verify is not None:\n",
    "            assert (sort_verify == sorted_camids).all(), \"INVALID SORT\"\n",
    "            compiled = np.concatenate((compiled, sorted_data), axis=1)\n",
    "        else:\n",
    "            sort_verify = sorted_camids\n",
    "            compiled = sorted_data\n",
    "            \n",
    "    return sort_verify, compiled\n",
    "\n",
    "def load_data(phenotype_folder, genome_folder, split_data_folder, chromosome, species, wing, color, out_dims_start_idx=0, out_dims=1):\n",
    "    \"\"\"\n",
    "    NOTE: There are missing camids from either pheno or geno type data\n",
    "    Seems to consistently be 1 missing from melpomene genotype\n",
    "    and 4 missing from erato phenotype\n",
    "    \"\"\"\n",
    "    pca_path = os.path.join(phenotype_folder, f\"{species}_{wing}_{color}\", \"data.csv\")\n",
    "    pca_df = pd.read_csv(pca_path)\n",
    "    pca_camids = pca_df.camid.to_numpy()\n",
    "\n",
    "    sort_idx = np.argsort(pca_camids)\n",
    "    pca_camids_sorted = pca_camids[sort_idx]\n",
    "    \n",
    "    pca_data = pca_df.iloc[:, (1 + out_dims_start_idx):(1 + out_dims_start_idx + out_dims)].to_numpy()\n",
    "    pca_data = pca_data[sort_idx]\n",
    "    \n",
    "    sorted_camids, compiled_data = collect_chromosome(genome_folder, species, chromosome)\n",
    "    conflict_camid_a = list(set.difference(set(pca_camids_sorted.tolist()), set(sorted_camids.tolist())))\n",
    "    conflict_camid_b = list(set.difference(set(sorted_camids.tolist()), set(pca_camids_sorted.tolist())))\n",
    "    conflict_camids = conflict_camid_a + conflict_camid_b\n",
    "    \n",
    "    idx = np.isin(sorted_camids, conflict_camids)\n",
    "    sorted_camids = sorted_camids[~idx]\n",
    "    compiled_data = compiled_data[~idx]\n",
    "    \n",
    "    idx = np.isin(pca_camids_sorted, conflict_camids)\n",
    "    pca_camids_sorted = pca_camids_sorted[~idx]\n",
    "    pca_data = pca_data[~idx]\n",
    "\n",
    "    assert (pca_camids_sorted == sorted_camids).all(), \"Invalid alignment\"\n",
    "\n",
    "    train_cams = np.load(os.path.join(split_data_folder, f\"{species}_train.npy\"))\n",
    "    val_cams = np.load(os.path.join(split_data_folder, f\"{species}_val.npy\"))\n",
    "    test_cams = np.load(os.path.join(split_data_folder, f\"{species}_test.npy\"))\n",
    "    \n",
    "    train_idx = np.isin(sorted_camids, train_cams)\n",
    "    train_pca_data = pca_data[train_idx]\n",
    "    train_geno_data = compiled_data[train_idx]\n",
    "    \n",
    "    val_idx = np.isin(sorted_camids, val_cams)\n",
    "    val_pca_data = pca_data[val_idx]\n",
    "    val_geno_data = compiled_data[val_idx]\n",
    "    \n",
    "    test_idx = np.isin(sorted_camids, test_cams)\n",
    "    test_pca_data = pca_data[test_idx]\n",
    "    test_geno_data = compiled_data[test_idx]\n",
    "    \n",
    "    return [train_geno_data, train_pca_data], [val_geno_data, val_pca_data], [test_geno_data, test_pca_data]\n",
    "\n",
    "def calc_pearson_correlation(model, dloader, device, index=0):\n",
    "    model.eval()\n",
    "    actual = []\n",
    "    predicted = []\n",
    "    for i, batch in enumerate(dloader):\n",
    "        model.zero_grad()\n",
    "        data, pca = batch\n",
    "        out = model(data.to(device))\n",
    "        actual.extend(pca[:, index].detach().cpu().numpy().tolist())\n",
    "        predicted.extend(out[:, index].detach().cpu().numpy().tolist())\n",
    "    return pearsonr(predicted, actual)\n",
    "\n",
    "def compile_attribution(attr):\n",
    "    attr, _ = torch.abs(attr).max(-1)\n",
    "    attr = attr[:, 0] # Only has 1 channel, just extract it\n",
    "    attr = attr.detach().cpu().numpy()\n",
    "    attr = np.median(np.abs(attr), 0) # median across batch???\n",
    "    #attr = attr.sum(0) # Sum across batch...Should we be summing here???\n",
    "    return attr\n",
    "\n",
    "def get_saliency_attr(m, dloader, device, target=0):\n",
    "    att_model = GuidedGradCam(m, m.last_block)\n",
    "    attr_total = None\n",
    "    for i, batch in enumerate(dloader):\n",
    "        m.zero_grad()\n",
    "        data, pca = batch\n",
    "        attr = att_model.attribute(data.to(device), target=target)\n",
    "        attr = compile_attribution(attr)\n",
    "        if attr_total is None:\n",
    "            attr_total = attr\n",
    "        else:\n",
    "            attr_total += attr\n",
    "\n",
    "    #attr_total = attr_total / np.linalg.norm(attr_total, ord=1) # Normalize\n",
    "    return attr_total\n",
    "\n",
    "def load_json_attributes(path):\n",
    "    with open(path, 'r') as f:\n",
    "        return json.load(f)\n",
    "    \n",
    "def save_json_attributes(path, data):\n",
    "    with open(path, 'w') as f:\n",
    "        json.dump(data, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "from gtp.dataloading.datasets import GTP_Dataset\n",
    "from gtp.models.net import SoyBeanNet\n",
    "\n",
    "def get_attribution_data(results_dir, num_chromosome, species, wing, color):\n",
    "    attributions = []\n",
    "    chromo_results = []\n",
    "    chromo_pvals = []\n",
    "    if REDO or not os.path.exists(os.path.join(results_dir, \"chromosome_summary_pearson.npy\")):\n",
    "        for chromosome in tqdm(range(1, num_chromosome+1), \"Extracting Chromosome Results\"):\n",
    "            train_data, val_data, test_data = load_data(PHENOTYPE_FOLDER, GENOME_FOLDER, DATA_SPLIT_FOLDER, \n",
    "                    chromosome, species, wing, color, out_dims_start_idx=0, out_dims=1)\n",
    "\n",
    "            train_dataset = GTP_Dataset(*train_data)\n",
    "            val_dataset = GTP_Dataset(*val_data)\n",
    "            test_dataset = GTP_Dataset(*test_data)\n",
    "\n",
    "            num_vcfs = train_data[0].shape[1]\n",
    "\n",
    "            train_dataloader = DataLoader(train_dataset, batch_size=BATCH_SIZE, num_workers=8, shuffle=True)\n",
    "            val_dataloader = DataLoader(val_dataset, batch_size=BATCH_SIZE, num_workers=8, shuffle=False)\n",
    "            test_dataloader = DataLoader(test_dataset, batch_size=BATCH_SIZE, num_workers=8, shuffle=False)\n",
    "\n",
    "            model = SoyBeanNet(window_size=num_vcfs, num_out_dims=1, insize=3, hidden_dim=10, drop_out_prob=0.75)\n",
    "            model_path = RESULTS_FOLDER / Path(f\"{species}_{wing}_{color}_chromosome_{chromosome}/model.pt\")\n",
    "            model.load_state_dict(torch.load(model_path))\n",
    "            model = model.to(DEVICE)\n",
    "            model.eval()\n",
    "\n",
    "            with torch.no_grad():\n",
    "                #tr_pearson_stat, _ = calc_pearson_correlation(model, train_dataloader, DEVICE)\n",
    "                val_pearson_stat, val_pv = calc_pearson_correlation(model, val_dataloader, DEVICE)\n",
    "                test_pearson_stat, test_pv = calc_pearson_correlation(model, test_dataloader, DEVICE)\n",
    "            val_att = get_saliency_attr(model, val_dataloader, DEVICE, target=target_index)\n",
    "            test_att = get_saliency_attr(model, test_dataloader, DEVICE, target=target_index)\n",
    "            attributions.append([val_att.tolist(), test_att.tolist()])\n",
    "            chromo_results.append([val_pearson_stat, test_pearson_stat])\n",
    "            chromo_pvals.append([f\"{val_pv:.2e}\", f\"{test_pv:.2e}\"])\n",
    "            \n",
    "            print(f\"Chromosome: {chromosome} - Pearson | (val) {val_pearson_stat} ({val_pv:.2e}) | (test) {test_pearson_stat} ({test_pv:.2e})\")\n",
    "            \n",
    "        np.save(os.path.join(results_dir, \"chromosome_summary_pearson.npy\"), chromo_results)\n",
    "        np.save(os.path.join(results_dir, \"chromosome_summary_pvals.npy\"), chromo_pvals)\n",
    "        save_json_attributes(os.path.join(results_dir, \"chromosome_attributions.json\"), attributions)\n",
    "    else:\n",
    "        chromo_results = np.load(os.path.join(results_dir, \"chromosome_summary_pearson.npy\"))\n",
    "        chromo_pvals = np.load(os.path.join(results_dir, \"chromosome_summary_pvals.npy\"))\n",
    "        attributions = load_json_attributes(os.path.join(results_dir, \"chromosome_attributions.json\"))\n",
    "    \n",
    "    return chromo_results, chromo_pvals, attributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap, LogNorm\n",
    "\n",
    "def plot_summary_results(results_dir, chromo_results, chromo_pvals, num_chromosome, species, wing, color):\n",
    "    bar_colors = [\n",
    "        (1, 0, 0),\n",
    "        (0, 0, 1)\n",
    "    ] # go from red to blue (significant to least significant)\n",
    "\n",
    "    pos = [0.0, 1.0]\n",
    "    bar_cmap = LinearSegmentedColormap.from_list('custom_colormap', \n",
    "                                                    list(zip(pos, bar_colors)))\n",
    "\n",
    "    fig, axs = plt.subplots(2, 1, figsize=(20, 14))\n",
    "    fig.suptitle(f\"{species} - {wing} - Phenotype:{color}\", fontsize=30)\n",
    "\n",
    "    chromo_results = np.array(chromo_results)\n",
    "    chromo_pvals = np.array(chromo_pvals)\n",
    "    chromo_labels = [f\"Chr. {chr}\" for chr in range(1, num_chromosome+1)]\n",
    "\n",
    "    for pi, phase in enumerate([\"val\", \"test\"]):\n",
    "\n",
    "        val_list = chromo_pvals[:, pi].astype(np.float64)\n",
    "        normed_vals = LogNorm()(val_list)\n",
    "        cur_colors = [bar_cmap(pv) for pv in normed_vals]\n",
    "        bar_plots = axs[pi].bar(chromo_labels, chromo_results[:, pi], color=cur_colors)\n",
    "        axs[pi].set_title(f\"{phase} Results\", fontsize=22)\n",
    "        axs[pi].set_ylim([0, 1])\n",
    "        axs[pi].set_xlabel(\"Chromosomes\", fontsize=22)\n",
    "        axs[pi].set_ylabel(\"Pearson Correlation Coefficient\", fontsize=22)\n",
    "\n",
    "        rects = axs[pi].patches\n",
    "\n",
    "        for rect, label in zip(rects, chromo_pvals[:, pi]):\n",
    "            label = \"\\ne\".join(label.split(\"e\"))\n",
    "            height = rect.get_height()\n",
    "            axs[pi].text(\n",
    "                rect.get_x() + rect.get_width() / 2, height + 0.01, \n",
    "                label, ha=\"center\", va=\"bottom\", rotation=0\n",
    "            )\n",
    "            \n",
    "        cbar = fig.colorbar(None, ax=axs[pi], cmap=bar_cmap)\n",
    "        cbar.ax.set_title(\"p-value\", fontsize=18)\n",
    "        cbar.ax.get_yaxis().set_ticks([1.0, 0.0])\n",
    "        cbar.ax.get_yaxis().set_ticklabels([val_list.max(), val_list.min()])\n",
    "\n",
    "    fig.tight_layout()\n",
    "\n",
    "    plt.savefig(os.path.join(results_dir / f\"genome_summary_{species}_{wing}_{color}.png\"))\n",
    "    if SHOW_PLOTS:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap, LogNorm\n",
    "\n",
    "def plot_detail_results(results_dir, chromo_results, chromo_pvals, attributions, num_chromosome, species, wing, color):\n",
    "    complete_data = {\n",
    "        \"val\" : [],\n",
    "        \"test\" : [],\n",
    "        \"val_start\" : [],\n",
    "        \"test_start\" : [],\n",
    "    }\n",
    "    bar_colors = [\n",
    "        (1, 0, 0),\n",
    "        (0, 0, 1)\n",
    "    ] # go from red to blue (significant to least significant)\n",
    "\n",
    "    pos = [0.0, 1.0]\n",
    "    bar_cmap = LinearSegmentedColormap.from_list('custom_colormap', \n",
    "                                                    list(zip(pos, bar_colors)))\n",
    "\n",
    "    fig, axs = plt.subplots(2, 1, figsize=(20, 14))\n",
    "    fig.suptitle(f\"{species} - {wing} - Phenotype:{color}\", fontsize=30)\n",
    "\n",
    "    chromo_labels = [f\"Chr. {chr}\" for chr in range(1, num_chromosome+1)]\n",
    "\n",
    "    start_pos = 0\n",
    "    alt_colors = [\"blue\", \"grey\"]\n",
    "    for pi, phase in enumerate([\"val\", \"test\"]):\n",
    "        chromo_pvals = np.array(chromo_pvals)\n",
    "        p_values = chromo_pvals[:, pi].astype(np.float64)\n",
    "        neg_log_p_vals = -np.log(p_values)\n",
    "        x_mid_points = []\n",
    "        for ci in tqdm(range(num_chromosome), desc=\"Plotting Attributions\"):\n",
    "            chromo_data = np.array(attributions[ci][pi])\n",
    "            num_pos = chromo_data.shape[0]\n",
    "            X = np.arange(num_pos) + start_pos\n",
    "            #adjust_val = p_values_normalized[ci]\n",
    "            Y = (chromo_data - chromo_data.min())\n",
    "            Y = Y / Y.max()\n",
    "            Y = Y * neg_log_p_vals[ci]\n",
    "            complete_data[phase].append(Y)\n",
    "            complete_data[phase+\"_start\"].append(start_pos)\n",
    "            start_pos = X[-1] + 1\n",
    "            bar_plots = axs[pi].scatter(X, Y, alpha=0.5, s=DOT_SIZE, color=alt_colors[ci % 2 == 1])\n",
    "            axs[pi].set_title(f\"{phase} Results\", fontsize=22)\n",
    "            axs[pi].set_xlabel(\"Chromosomes/SNP Position\", fontsize=22)\n",
    "            axs[pi].set_ylabel(\"attribution * -log(chromosome p-value)\", fontsize=22)\n",
    "            mid_point = X[X.shape[0]//2]\n",
    "            x_mid_points.append(mid_point)\n",
    "            axs[pi].text(\n",
    "                mid_point, 0, \n",
    "                f\"{p_values[ci]:.1e}\", ha=\"center\", va=\"bottom\", rotation=0, color=\"red\"\n",
    "            )\n",
    "        axs[pi].set_xticks(x_mid_points)\n",
    "        axs[pi].set_xticklabels(chromo_labels)\n",
    "\n",
    "    fig.tight_layout()\n",
    "\n",
    "    plt.savefig(os.path.join(results_dir, f\"genome_detail_{species}_{wing}_{color}.png\"))\n",
    "    if SHOW_PLOTS:\n",
    "        plt.show()\n",
    "        \n",
    "    np.savetxt(os.path.join(results_dir, f\"val_plot_data.csv\"), np.concatenate(complete_data[\"val\"], axis=0), delimiter=\",\")\n",
    "    np.savetxt(os.path.join(results_dir, f\"val_plot_data_chromosome_starts.csv\"), np.array(complete_data[\"val_start\"]), delimiter=\",\")\n",
    "    np.savetxt(os.path.join(results_dir, f\"test_plot_data.csv\"), np.concatenate(complete_data[\"test\"], axis=0), delimiter=\",\")\n",
    "    np.savetxt(os.path.join(results_dir, f\"test_plot_data_chromosome_starts.csv\"), np.array(complete_data[\"test_start\"]), delimiter=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting Chromosome Results:   0%|          | 0/21 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/carlyn.1/miniconda3/envs/dna/lib/python3.12/site-packages/torch/nn/modules/conv.py:456: UserWarning: Using padding='same' with even kernel lengths and odd dilation may require a zero-padded copy of the input be created (Triggered internally at ../aten/src/ATen/native/Convolution.cpp:1031.)\n",
      "  return F.conv2d(input, weight, bias, self.stride,\n",
      "/home/carlyn.1/miniconda3/envs/dna/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.\n",
      "  warnings.warn(\n",
      "/home/carlyn.1/miniconda3/envs/dna/lib/python3.12/site-packages/captum/attr/_core/guided_backprop_deconvnet.py:64: UserWarning: Setting backward hooks on ReLU activations.The hooks will be removed after the attribution is finished\n",
      "  warnings.warn(\n",
      "Extracting Chromosome Results:   0%|          | 0/21 [00:33<?, ?it/s]\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "cannot access local variable 'attributions' where it is not associated with a value",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome copy.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=14'>15</a>\u001b[0m results_dir \u001b[39m=\u001b[39m PLOT_RESULTS_FOLDER \u001b[39m/\u001b[39m \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mspecies\u001b[39m}\u001b[39;00m\u001b[39m_\u001b[39m\u001b[39m{\u001b[39;00mwing\u001b[39m}\u001b[39;00m\u001b[39m_\u001b[39m\u001b[39m{\u001b[39;00mcolor\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=15'>16</a>\u001b[0m os\u001b[39m.\u001b[39mmakedirs(results_dir, exist_ok\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m---> <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=17'>18</a>\u001b[0m chromo_results, chromo_pvals, attributions \u001b[39m=\u001b[39m get_attribution_data(results_dir, num_chromosome, species, wing, color)\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=18'>19</a>\u001b[0m plot_summary_results(results_dir, chromo_results, chromo_pvals, num_chromosome, species, wing, color)\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=19'>20</a>\u001b[0m plot_detail_results(results_dir, chromo_results, chromo_pvals, attributions, num_chromosome, species, wing, color)\n",
      "\u001b[1;32m/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome copy.ipynb Cell 6\u001b[0m line \u001b[0;36m3\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=35'>36</a>\u001b[0m val_att \u001b[39m=\u001b[39m get_saliency_attr(model, val_dataloader, DEVICE, target\u001b[39m=\u001b[39mtarget_index)\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=36'>37</a>\u001b[0m test_att \u001b[39m=\u001b[39m get_saliency_attr(model, test_dataloader, DEVICE, target\u001b[39m=\u001b[39mtarget_index)\n\u001b[0;32m---> <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=37'>38</a>\u001b[0m attributions\u001b[39m.\u001b[39mappend([val_att\u001b[39m.\u001b[39mtolist(), test_att\u001b[39m.\u001b[39mtolist()])\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=38'>39</a>\u001b[0m chromo_results\u001b[39m.\u001b[39mappend([val_pearson_stat, test_pearson_stat])\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2Bcse-cnc196909s.coeit.osu.edu/home/carlyn.1/dna-trait-analysis/notebooks/plot_whole_genome%20copy.ipynb#W5sdnNjb2RlLXJlbW90ZQ%3D%3D?line=39'>40</a>\u001b[0m chromo_pvals\u001b[39m.\u001b[39mappend([\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mval_pv\u001b[39m:\u001b[39;00m\u001b[39m.2e\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m, \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mtest_pv\u001b[39m:\u001b[39;00m\u001b[39m.2e\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m])\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: cannot access local variable 'attributions' where it is not associated with a value"
     ]
    }
   ],
   "source": [
    "species = \"erato\"\n",
    "wing = \"forewings\"\n",
    "color = \"total\"\n",
    "target_index = 0\n",
    "chromo_results = []\n",
    "chromo_pvals = []\n",
    "attributions = []\n",
    "num_chromosome = 21\n",
    "DOT_SIZE = 10\n",
    "SHOW_PLOTS = False\n",
    "\n",
    "for species in [\"melpomene\"]:\n",
    "    for wing in [\"forewings\"]:\n",
    "        for color in [\"color_3\", \"total\"]:\n",
    "            results_dir = PLOT_RESULTS_FOLDER / f\"{species}_{wing}_{color}\"\n",
    "            os.makedirs(results_dir, exist_ok=True)\n",
    "\n",
    "            chromo_results, chromo_pvals, attributions = get_attribution_data(results_dir, num_chromosome, species, wing, color)\n",
    "            plot_summary_results(results_dir, chromo_results, chromo_pvals, num_chromosome, species, wing, color)\n",
    "            plot_detail_results(results_dir, chromo_results, chromo_pvals, attributions, num_chromosome, species, wing, color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1763528/282410606.py:7: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed two minor releases later. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.\n",
      "  cmap = cm.get_cmap('PuBu_r')\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiMAAAGdCAYAAADAAnMpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAcrUlEQVR4nO3df3DUd5348RcJzYZaoD84EsD0gloF+gMoOTIpOuqYa64yOIyjYu0VJlZu9JI5IKNXaAu5Hpa0VThOi82Bov4hgnq26tHDwVyp02taWmhu7PcKvV7bg6mXAFMlNLWkJvv9w2k6OQJl08Q3mz4eM/tHPvt57+e1KzbP+exns6Oy2Ww2AAASKUg9AADw9iZGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgqdGpBzgbvb298etf/zrGjh0bo0aNSj0OAHAWstlsnDhxIiZPnhwFBac//5EXMfLrX/86ysrKUo8BAAzC4cOH453vfOdp78+LGBk7dmxE/OHJjBs3LvE0AMDZ6OzsjLKysr7f46eTFzHy+lsz48aNEyMAkGfe7BILF7ACAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAIKmcY+SXv/xlLFiwICZPnhyjRo2K+++//03X7NmzJ66++urIZDLxnve8J77zne8MYlQAYCTKOUa6urpi5syZsWnTprPa//nnn4/58+fHhz/84Whra4vly5fH5z73ufj5z3+e87AAwMiT8xflXXfddXHddded9f7Nzc0xderUWL9+fURETJ8+PR5++OH4h3/4h6ipqcn18ADACDPs14y0trZGdXV1v201NTXR2tp62jUnT56Mzs7OfjcAYGTK+cxIrtrb26OkpKTftpKSkujs7Izf/e53MWbMmFPWNDU1xe233z7co0VERGHV9X+U47yZntbv57zmvEV3DsMkuXltx8qc1xTVbx+GSXLTfc+nc17zvnW/GIZJcnfwluo33+n/uP7bjw3DJLn5fm1lzmu++rP/NwyT5O6LCy7Pec3Of39h6AfJ0fx55TmvOfD0kaEfJEfTpk/Mec1vjrw8DJPk7qKJF+S+6NXfD/0guSoe9hw4o3Py0zSrVq2K48eP990OHz6ceiQAYJgMewqVlpZGR0dHv20dHR0xbty4Ac+KRERkMpnIZDLDPRoAcA4Y9jMjVVVV0dLS0m/b7t27o6qqargPDQDkgZxj5OWXX462trZoa2uLiD98dLetrS0OHToUEX94i2Xx4sV9+3/+85+P5557Lv72b/82Dhw4EN/4xjfiBz/4QaxYsWJongEAkNdyjpEnnngiZs+eHbNnz46IiIaGhpg9e3asWbMmIiL+93//ty9MIiKmTp0aO3fujN27d8fMmTNj/fr18c1vftPHegGAiBjENSMf+tCHIpvNnvb+gf666oc+9KF48skncz0UAPA2cE5+mgYAePsQIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEkNKkY2bdoU5eXlUVxcHJWVlbF3794z7r9x48Z43/veF2PGjImysrJYsWJFvPrqq4MaGAAYWXKOkR07dkRDQ0M0NjbG/v37Y+bMmVFTUxNHjhwZcP9t27bFypUro7GxMZ5++un41re+FTt27IhbbrnlLQ8PAOS/nGNkw4YNsXTp0qitrY0ZM2ZEc3NznH/++bF169YB93/kkUdi3rx58ZnPfCbKy8vj2muvjeuvv/5Nz6YAAG8POcVId3d37Nu3L6qrq994gIKCqK6ujtbW1gHXXHPNNbFv376++HjuuefigQceiI9+9KOnPc7Jkyejs7Oz3w0AGJlG57LzsWPHoqenJ0pKSvptLykpiQMHDgy45jOf+UwcO3Ys3v/+90c2m43f//738fnPf/6Mb9M0NTXF7bffnstoAECeGvZP0+zZsyfWrVsX3/jGN2L//v3x4x//OHbu3Blr16497ZpVq1bF8ePH+26HDx8e7jEBgERyOjMyYcKEKCwsjI6Ojn7bOzo6orS0dMA1q1evjhtvvDE+97nPRUTElVdeGV1dXfFXf/VXceutt0ZBwak9lMlkIpPJ5DIaAJCncjozUlRUFHPmzImWlpa+bb29vdHS0hJVVVUDrnnllVdOCY7CwsKIiMhms7nOCwCMMDmdGYmIaGhoiCVLlkRFRUXMnTs3Nm7cGF1dXVFbWxsREYsXL44pU6ZEU1NTREQsWLAgNmzYELNnz47Kysp49tlnY/Xq1bFgwYK+KAEA3r5yjpFFixbF0aNHY82aNdHe3h6zZs2KXbt29V3UeujQoX5nQm677bYYNWpU3HbbbfHiiy/Gn/zJn8SCBQvijjvuGLpnAQDkrZxjJCKivr4+6uvrB7xvz549/Q8wenQ0NjZGY2PjYA4FAIxwvpsGAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgqUHFyKZNm6K8vDyKi4ujsrIy9u7de8b9f/vb30ZdXV1MmjQpMplMvPe9740HHnhgUAMDACPL6FwX7NixIxoaGqK5uTkqKytj48aNUVNTEwcPHoyJEyeesn93d3f8+Z//eUycODF+9KMfxZQpU+J//ud/4sILLxyK+QGAPJdzjGzYsCGWLl0atbW1ERHR3NwcO3fujK1bt8bKlStP2X/r1q3x0ksvxSOPPBLnnXdeRESUl5e/takBgBEjp7dpuru7Y9++fVFdXf3GAxQURHV1dbS2tg645qc//WlUVVVFXV1dlJSUxBVXXBHr1q2Lnp6e0x7n5MmT0dnZ2e8GAIxMOcXIsWPHoqenJ0pKSvptLykpifb29gHXPPfcc/GjH/0oenp64oEHHojVq1fH+vXr48tf/vJpj9PU1BTjx4/vu5WVleUyJgCQR4b90zS9vb0xceLE2Lx5c8yZMycWLVoUt956azQ3N592zapVq+L48eN9t8OHDw/3mABAIjldMzJhwoQoLCyMjo6Ofts7OjqitLR0wDWTJk2K8847LwoLC/u2TZ8+Pdrb26O7uzuKiopOWZPJZCKTyeQyGgCQp3I6M1JUVBRz5syJlpaWvm29vb3R0tISVVVVA66ZN29ePPvss9Hb29u37ZlnnolJkyYNGCIAwNtLzm/TNDQ0xJYtW+K73/1uPP300/GFL3whurq6+j5ds3jx4li1alXf/l/4whfipZdeimXLlsUzzzwTO3fujHXr1kVdXd3QPQsAIG/l/NHeRYsWxdGjR2PNmjXR3t4es2bNil27dvVd1Hro0KEoKHijccrKyuLnP/95rFixIq666qqYMmVKLFu2LG6++eahexYAQN7KOUYiIurr66O+vn7A+/bs2XPKtqqqqnj00UcHcygAYITz3TQAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJDSpGNm3aFOXl5VFcXByVlZWxd+/es1q3ffv2GDVqVCxcuHAwhwUARqCcY2THjh3R0NAQjY2NsX///pg5c2bU1NTEkSNHzrjuhRdeiC9+8YvxgQ98YNDDAgAjT84xsmHDhli6dGnU1tbGjBkzorm5Oc4///zYunXradf09PTEDTfcELfffnu8613veksDAwAjS04x0t3dHfv27Yvq6uo3HqCgIKqrq6O1tfW06/7+7/8+Jk6cGDfddNNZHefkyZPR2dnZ7wYAjEw5xcixY8eip6cnSkpK+m0vKSmJ9vb2Adc8/PDD8a1vfSu2bNly1sdpamqK8ePH993KyspyGRMAyCPD+mmaEydOxI033hhbtmyJCRMmnPW6VatWxfHjx/tuhw8fHsYpAYCURuey84QJE6KwsDA6Ojr6be/o6IjS0tJT9v/v//7veOGFF2LBggV923p7e/9w4NGj4+DBg/Hud7/7lHWZTCYymUwuowEAeSqnMyNFRUUxZ86caGlp6dvW29sbLS0tUVVVdcr+06ZNi1/96lfR1tbWd/vYxz4WH/7wh6Otrc3bLwBAbmdGIiIaGhpiyZIlUVFREXPnzo2NGzdGV1dX1NbWRkTE4sWLY8qUKdHU1BTFxcVxxRVX9Ft/4YUXRkScsh0AeHvKOUYWLVoUR48ejTVr1kR7e3vMmjUrdu3a1XdR66FDh6KgwB92BQDOTs4xEhFRX18f9fX1A963Z8+eM679zne+M5hDAgAjlFMYAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACCpQcXIpk2bory8PIqLi6OysjL27t172n23bNkSH/jAB+Kiiy6Kiy66KKqrq8+4PwDw9pJzjOzYsSMaGhqisbEx9u/fHzNnzoyampo4cuTIgPvv2bMnrr/++njwwQejtbU1ysrK4tprr40XX3zxLQ8PAOS/nGNkw4YNsXTp0qitrY0ZM2ZEc3NznH/++bF169YB9//e974Xf/3Xfx2zZs2KadOmxTe/+c3o7e2NlpaWtzw8AJD/coqR7u7u2LdvX1RXV7/xAAUFUV1dHa2trWf1GK+88kq89tprcfHFF592n5MnT0ZnZ2e/GwAwMuUUI8eOHYuenp4oKSnpt72kpCTa29vP6jFuvvnmmDx5cr+g+b+amppi/PjxfbeysrJcxgQA8sgf9dM0d955Z2zfvj3uu+++KC4uPu1+q1atiuPHj/fdDh8+/EecEgD4Yxqdy84TJkyIwsLC6Ojo6Le9o6MjSktLz7j2q1/9atx5553xi1/8Iq666qoz7pvJZCKTyeQyGgCQp3I6M1JUVBRz5szpd/Hp6xejVlVVnXbd3XffHWvXro1du3ZFRUXF4KcFAEacnM6MREQ0NDTEkiVLoqKiIubOnRsbN26Mrq6uqK2tjYiIxYsXx5QpU6KpqSkiIu66665Ys2ZNbNu2LcrLy/uuLbngggviggsuGMKnAgDko5xjZNGiRXH06NFYs2ZNtLe3x6xZs2LXrl19F7UeOnQoCgreOOFy7733Rnd3d3ziE5/o9ziNjY3xd3/3d29tegAg7+UcIxER9fX1UV9fP+B9e/bs6ffzCy+8MJhDAABvE76bBgBISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAIKlBxcimTZuivLw8iouLo7KyMvbu3XvG/X/4wx/GtGnTori4OK688sp44IEHBjUsADDy5BwjO3bsiIaGhmhsbIz9+/fHzJkzo6amJo4cOTLg/o888khcf/31cdNNN8WTTz4ZCxcujIULF8ZTTz31locHAPJfzjGyYcOGWLp0adTW1saMGTOiubk5zj///Ni6deuA+//jP/5j/MVf/EV86UtfiunTp8fatWvj6quvjnvuuectDw8A5L/Ruezc3d0d+/bti1WrVvVtKygoiOrq6mhtbR1wTWtrazQ0NPTbVlNTE/fff/9pj3Py5Mk4efJk38/Hjx+PiIjOzs5cxj0r2d+/NuSPORiDeW7Z114dhklyM6i5u18ZhklyM5i5e17tGoZJcjeY2V/7XfrZBzP3q6+8PAyT5G4ws7/SdWIYJsnNYOZ++eVzYe7i3NecODf+rRQW9+a+6NXfD/0guerOKQfO2uv/BrPZ7Bn3y+nox44di56enigpKem3vaSkJA4cODDgmvb29gH3b29vP+1xmpqa4vbbbz9le1lZWS7j5pXx4/859QiDMv6+U/93ygfjt9yUeoRBG7829QSD8891qScYvNWpB4A8d+LEiRg/fvxp7x+eFHqLVq1a1e9sSm9vb7z00ktxySWXxKhRoxJOdqrOzs4oKyuLw4cPx7hx41KPk9e8lkPL6zl0vJZDy+s5dM711zKbzcaJEydi8uTJZ9wvpxiZMGFCFBYWRkdHR7/tHR0dUVpaOuCa0tLSnPaPiMhkMpHJZPptu/DCC3MZ9Y9u3Lhx5+Q/hHzktRxaXs+h47UcWl7PoXMuv5ZnOiPyupwuYC0qKoo5c+ZES0tL37be3t5oaWmJqqqqAddUVVX12z8iYvfu3afdHwB4e8n5bZqGhoZYsmRJVFRUxNy5c2Pjxo3R1dUVtbW1ERGxePHimDJlSjQ1NUVExLJly+KDH/xgrF+/PubPnx/bt2+PJ554IjZv3jy0zwQAyEs5x8iiRYvi6NGjsWbNmmhvb49Zs2bFrl27+i5SPXToUBQUvHHC5Zprrolt27bFbbfdFrfccktcdtllcf/998cVV1wxdM8ioUwmE42Njae8rUTuvJZDy+s5dLyWQ8vrOXRGyms5Kvtmn7cBABhGvpsGAEhKjAAASYkRACApMQIAJCVG3oJNmzZFeXl5FBcXR2VlZezduzf1SHmpqakp/uzP/izGjh0bEydOjIULF8bBgwdTjzUi3HnnnTFq1KhYvnx56lHy1osvvhh/+Zd/GZdcckmMGTMmrrzyynjiiSdSj5WXenp6YvXq1TF16tQYM2ZMvPvd7461a9e+6feWEPHLX/4yFixYEJMnT45Ro0ad8v1u2Ww21qxZE5MmTYoxY8ZEdXV1/Nd//VeaYQdBjAzSjh07oqGhIRobG2P//v0xc+bMqKmpiSNHjqQeLe889NBDUVdXF48++mjs3r07Xnvttbj22mujqyv9l7vls8cffzz+6Z/+Ka666qrUo+St3/zmNzFv3rw477zz4l//9V/jP//zP2P9+vVx0UUXpR4tL911111x7733xj333BNPP/103HXXXXH33XfH17/+9dSjnfO6urpi5syZsWnTpgHvv/vuu+NrX/taNDc3x2OPPRbveMc7oqamJl59Nf0Xqp6VLIMyd+7cbF1dXd/PPT092cmTJ2ebmpoSTjUyHDlyJBsR2Yceeij1KHnrxIkT2csuuyy7e/fu7Ac/+MHssmXLUo+Ul26++ebs+9///tRjjBjz58/Pfvazn+237eMf/3j2hhtuSDRRfoqI7H333df3c29vb7a0tDT7la98pW/bb3/722wmk8l+//vfTzBh7pwZGYTu7u7Yt29fVFdX920rKCiI6urqaG1tTTjZyHD8+PGIiLj44osTT5K/6urqYv78+f3+jZK7n/70p1FRURGf/OQnY+LEiTF79uzYsmVL6rHy1jXXXBMtLS3xzDPPRETEf/zHf8TDDz8c1113XeLJ8tvzzz8f7e3t/f7/Pn78+KisrMyb30nn5Lf2nuuOHTsWPT09fX919nUlJSVx4MCBRFONDL29vbF8+fKYN2/eiPkrvX9s27dvj/3798fjjz+eepS899xzz8W9994bDQ0Nccstt8Tjjz8ef/M3fxNFRUWxZMmS1OPlnZUrV0ZnZ2dMmzYtCgsLo6enJ+6444644YYbUo+W19rb2yMiBvyd9Pp95zoxwjmlrq4unnrqqXj44YdTj5KXDh8+HMuWLYvdu3dHcXFx6nHyXm9vb1RUVMS6desiImL27Nnx1FNPRXNzsxgZhB/84Afxve99L7Zt2xaXX355tLW1xfLly2Py5Mlez7c5b9MMwoQJE6KwsDA6Ojr6be/o6IjS0tJEU+W/+vr6+Jd/+Zd48MEH453vfGfqcfLSvn374siRI3H11VfH6NGjY/To0fHQQw/F1772tRg9enT09PSkHjGvTJo0KWbMmNFv2/Tp0+PQoUOJJspvX/rSl2LlypXx6U9/Oq688sq48cYbY8WKFX1frMrgvP57J59/J4mRQSgqKoo5c+ZES0tL37be3t5oaWmJqqqqhJPlp2w2G/X19XHffffFv/3bv8XUqVNTj5S3PvKRj8SvfvWraGtr67tVVFTEDTfcEG1tbVFYWJh6xLwyb968Uz5m/swzz8Sf/umfJpoov73yyiv9vkg1IqKwsDB6e3sTTTQyTJ06NUpLS/v9Turs7IzHHnssb34neZtmkBoaGmLJkiVRUVERc+fOjY0bN0ZXV1fU1tamHi3v1NXVxbZt2+InP/lJjB07tu89zvHjx8eYMWMST5dfxo4de8q1Nu94xzvikksucQ3OIKxYsSKuueaaWLduXXzqU5+KvXv3xubNm2Pz5s2pR8tLCxYsiDvuuCMuvfTSuPzyy+PJJ5+MDRs2xGc/+9nUo53zXn755Xj22Wf7fn7++eejra0tLr744rj00ktj+fLl8eUvfzkuu+yymDp1aqxevTomT54cCxcuTDd0LlJ/nCefff3rX89eeuml2aKiouzcuXOzjz76aOqR8lJEDHj79re/nXq0EcFHe9+an/3sZ9krrrgim8lkstOmTctu3rw59Uh5q7OzM7ts2bLspZdemi0uLs6+613vyt56663ZkydPph7tnPfggw8O+N/JJUuWZLPZP3y8d/Xq1dmSkpJsJpPJfuQjH8kePHgw7dA5GJXN+tN3AEA6rhkBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEn9f8/BSBcubKhxAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\"\\n# A low hump with a spike coming out of the top right.  Needs to have\\n# z/colour axis on a log scale, so we see both hump and spike. A linear\\n# scale only shows the spike.\\nZ1 = np.exp(-X**2 - Y**2)\\nZ2 = np.exp(-(X * 10)**2 - (Y * 10)**2)\\nZ = Z1 + 50 * Z2\\n\\nfig, ax = plt.subplots(2, 1)\\n\\npcm = ax[0].pcolor(X, Y, Z,\\n                   norm=colors.LogNorm(vmin=Z.min(), vmax=Z.max()),\\n                   cmap='PuBu_r', shading='auto')\\nfig.colorbar(pcm, ax=ax[0], extend='max')\\n\\npcm = ax[1].pcolor(X, Y, Z, cmap='PuBu_r', shading='auto')\\nfig.colorbar(pcm, ax=ax[1], extend='max')\\nplt.show()\\n\""
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from matplotlib import cm\n",
    "import matplotlib.cbook as cbook\n",
    "import matplotlib.colors as colors\n",
    "\n",
    "cmap = cm.get_cmap('PuBu_r')\n",
    "log_norm = colors.LogNorm()\n",
    "\n",
    "X = np.linspace(0, 10, 10)\n",
    "Y = np.logspace(0, 1e-6, 10, base=1e-20)[::-1]\n",
    "\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "ax.bar(X, Y, color=cmap(log_norm(Y)))\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "# A low hump with a spike coming out of the top right.  Needs to have\n",
    "# z/colour axis on a log scale, so we see both hump and spike. A linear\n",
    "# scale only shows the spike.\n",
    "Z1 = np.exp(-X**2 - Y**2)\n",
    "Z2 = np.exp(-(X * 10)**2 - (Y * 10)**2)\n",
    "Z = Z1 + 50 * Z2\n",
    "\n",
    "fig, ax = plt.subplots(2, 1)\n",
    "\n",
    "pcm = ax[0].pcolor(X, Y, Z,\n",
    "                   norm=colors.LogNorm(vmin=Z.min(), vmax=Z.max()),\n",
    "                   cmap='PuBu_r', shading='auto')\n",
    "fig.colorbar(pcm, ax=ax[0], extend='max')\n",
    "\n",
    "pcm = ax[1].pcolor(X, Y, Z, cmap='PuBu_r', shading='auto')\n",
    "fig.colorbar(pcm, ax=ax[1], extend='max')\n",
    "plt.show()\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
