{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Scaffold",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Start",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "End",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Length",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Intervals",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Direction",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "info",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "94d4ea2e-ceb3-402e-89aa-57ee432d6994",
       "rows": [
        [
         "0",
         "Herato1001",
         "wnta",
         "gene",
         "4626226",
         "4707387",
         "81161",
         "1",
         "forward",
         "vanbellemg concaha"
        ],
        [
         "1",
         "Herato1001",
         "Sd1",
         "phenotype",
         "4637657",
         "4637727",
         "70",
         "1",
         "forward",
         "vanbellemg concaha"
        ],
        [
         "2",
         "Herato1001",
         "Sd2",
         "phenotype",
         "4639853",
         "4641535",
         "1682",
         "1",
         "forward",
         "vanbellemg concaha"
        ],
        [
         "3",
         "Herato1001",
         "St",
         "phenotype",
         "4657452",
         "4658207",
         "755",
         "1",
         "forward",
         "vanbellemg concaha"
        ],
        [
         "4",
         "Herato1001",
         "Ly1",
         "phenotype",
         "4666909",
         "4670474",
         "3565",
         "1",
         "forward",
         "vanbellemg concaha"
        ]
       ],
       "shape": {
        "columns": 9,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Scaffold</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length</th>\n",
       "      <th>Intervals</th>\n",
       "      <th>Direction</th>\n",
       "      <th>info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>wnta</td>\n",
       "      <td>gene</td>\n",
       "      <td>4626226</td>\n",
       "      <td>4707387</td>\n",
       "      <td>81161</td>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>vanbellemg concaha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>Sd1</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>4637657</td>\n",
       "      <td>4637727</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>vanbellemg concaha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>Sd2</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>4639853</td>\n",
       "      <td>4641535</td>\n",
       "      <td>1682</td>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>vanbellemg concaha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>St</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>4657452</td>\n",
       "      <td>4658207</td>\n",
       "      <td>755</td>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>vanbellemg concaha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>Ly1</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>4666909</td>\n",
       "      <td>4670474</td>\n",
       "      <td>3565</td>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>vanbellemg concaha</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Scaffold  Name       Type    Start      End  Length  Intervals Direction  \\\n",
       "0  Herato1001  wnta       gene  4626226  4707387   81161          1   forward   \n",
       "1  Herato1001   Sd1  phenotype  4637657  4637727      70          1   forward   \n",
       "2  Herato1001   Sd2  phenotype  4639853  4641535    1682          1   forward   \n",
       "3  Herato1001    St  phenotype  4657452  4658207     755          1   forward   \n",
       "4  Herato1001   Ly1  phenotype  4666909  4670474    3565          1   forward   \n",
       "\n",
       "                 info  \n",
       "0  vanbellemg concaha  \n",
       "1  vanbellemg concaha  \n",
       "2  vanbellemg concaha  \n",
       "3  vanbellemg concaha  \n",
       "4  vanbellemg concaha  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "gwas_points_file_path = Path(\"/local/scratch/carlyn.1/dna/metadata/gwas_Herato1001_4570000_4850000.txt\")\n",
    "gwas_df = pd.read_csv(gwas_points_file_path, delimiter=\"\\t\")\n",
    "\n",
    "annotation_file_path = Path(\"/local/scratch/carlyn.1/dna/metadata/wnta_annotation_mel2x5_eratodem1.xlsx\")\n",
    "anno_df = pd.read_excel(annotation_file_path,  sheet_name=\"Complete_wnta_annt_demopho\")\n",
    "anno_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "chr",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "rs",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ps",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "n_miss",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "allele1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "allele0",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "af",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "beta",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "se",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "logl_H1",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "l_remle",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "l_mle",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "p_wald",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "p_lrt",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "p_score",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "neg_log10_p",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "72158b08-f1fa-4660-8c6a-b25c35d253fb",
       "rows": [
        [
         "0",
         "Herato1001",
         ".",
         "4639773",
         "0",
         "A",
         "C",
         "0.303",
         "-56.16131",
         "1.913198",
         "-2210.711",
         "6808.26",
         "100000.0",
         "1.912446e-108",
         "5.500456e-109",
         "3.923419e-53",
         "107.7184108187799"
        ],
        [
         "1",
         "Herato1001",
         ".",
         "4639853",
         "0",
         "C",
         "G",
         "0.298",
         "-56.35173",
         "1.929564",
         "-2212.244",
         "100000.0",
         "100000.0",
         "9.120289e-108",
         "2.555516e-108",
         "5.462232e-53",
         "107.03999139970703"
        ],
        [
         "2",
         "Herato1001",
         ".",
         "4639936",
         "0",
         "C",
         "T",
         "0.299",
         "-56.80083",
         "1.949337",
         "-2212.931",
         "100000.0",
         "100000.0",
         "1.809534e-107",
         "5.08816e-108",
         "6.341158e-53",
         "106.74243325236031"
        ],
        [
         "3",
         "Herato1001",
         ".",
         "4640008",
         "0",
         "A",
         "T",
         "0.305",
         "-56.54915",
         "1.880458",
         "-2204.094",
         "294.4706",
         "100000.0",
         "1.141958e-111",
         "7.255491000000002e-112",
         "9.685145e-54",
         "110.94234986868533"
        ],
        [
         "4",
         "Herato1001",
         ".",
         "4640096",
         "0",
         "C",
         "A",
         "0.305",
         "-57.67782",
         "1.856515",
         "-2194.273",
         "199.9867",
         "100000.0",
         "4.196162e-116",
         "3.865583e-116",
         "1.322778e-54",
         "115.37714775347467"
        ]
       ],
       "shape": {
        "columns": 16,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rs</th>\n",
       "      <th>ps</th>\n",
       "      <th>n_miss</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele0</th>\n",
       "      <th>af</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>logl_H1</th>\n",
       "      <th>l_remle</th>\n",
       "      <th>l_mle</th>\n",
       "      <th>p_wald</th>\n",
       "      <th>p_lrt</th>\n",
       "      <th>p_score</th>\n",
       "      <th>neg_log10_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>.</td>\n",
       "      <td>4639773</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.303</td>\n",
       "      <td>-56.16131</td>\n",
       "      <td>1.913198</td>\n",
       "      <td>-2210.711</td>\n",
       "      <td>6808.2600</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>1.912446e-108</td>\n",
       "      <td>5.500456e-109</td>\n",
       "      <td>3.923419e-53</td>\n",
       "      <td>107.718411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>.</td>\n",
       "      <td>4639853</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.298</td>\n",
       "      <td>-56.35173</td>\n",
       "      <td>1.929564</td>\n",
       "      <td>-2212.244</td>\n",
       "      <td>100000.0000</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>9.120289e-108</td>\n",
       "      <td>2.555516e-108</td>\n",
       "      <td>5.462232e-53</td>\n",
       "      <td>107.039991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>.</td>\n",
       "      <td>4639936</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.299</td>\n",
       "      <td>-56.80083</td>\n",
       "      <td>1.949337</td>\n",
       "      <td>-2212.931</td>\n",
       "      <td>100000.0000</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>1.809534e-107</td>\n",
       "      <td>5.088160e-108</td>\n",
       "      <td>6.341158e-53</td>\n",
       "      <td>106.742433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>.</td>\n",
       "      <td>4640008</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.305</td>\n",
       "      <td>-56.54915</td>\n",
       "      <td>1.880458</td>\n",
       "      <td>-2204.094</td>\n",
       "      <td>294.4706</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>1.141958e-111</td>\n",
       "      <td>7.255491e-112</td>\n",
       "      <td>9.685145e-54</td>\n",
       "      <td>110.942350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Herato1001</td>\n",
       "      <td>.</td>\n",
       "      <td>4640096</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.305</td>\n",
       "      <td>-57.67782</td>\n",
       "      <td>1.856515</td>\n",
       "      <td>-2194.273</td>\n",
       "      <td>199.9867</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>4.196162e-116</td>\n",
       "      <td>3.865583e-116</td>\n",
       "      <td>1.322778e-54</td>\n",
       "      <td>115.377148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr rs       ps  n_miss allele1 allele0     af      beta        se  \\\n",
       "0  Herato1001  .  4639773       0       A       C  0.303 -56.16131  1.913198   \n",
       "1  Herato1001  .  4639853       0       C       G  0.298 -56.35173  1.929564   \n",
       "2  Herato1001  .  4639936       0       C       T  0.299 -56.80083  1.949337   \n",
       "3  Herato1001  .  4640008       0       A       T  0.305 -56.54915  1.880458   \n",
       "4  Herato1001  .  4640096       0       C       A  0.305 -57.67782  1.856515   \n",
       "\n",
       "    logl_H1      l_remle     l_mle         p_wald          p_lrt  \\\n",
       "0 -2210.711    6808.2600  100000.0  1.912446e-108  5.500456e-109   \n",
       "1 -2212.244  100000.0000  100000.0  9.120289e-108  2.555516e-108   \n",
       "2 -2212.931  100000.0000  100000.0  1.809534e-107  5.088160e-108   \n",
       "3 -2204.094     294.4706  100000.0  1.141958e-111  7.255491e-112   \n",
       "4 -2194.273     199.9867  100000.0  4.196162e-116  3.865583e-116   \n",
       "\n",
       "        p_score  neg_log10_p  \n",
       "0  3.923419e-53   107.718411  \n",
       "1  5.462232e-53   107.039991  \n",
       "2  6.341158e-53   106.742433  \n",
       "3  9.685145e-54   110.942350  \n",
       "4  1.322778e-54   115.377148  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_DIR = Path(\"/local/scratch/carlyn.1/dna/training_output/pca_10/\")\n",
    "DNA_ROOT_DIR = Path(\"/local/scratch/carlyn.1/dna/processed/genotypes/genome/\")\n",
    "EDIT_OPTIONS = [\"AA\", \"Aa/aA\", \"aa\", \"zero-out\"]\n",
    "PHASE_NAMES = [\"training\", \"validation\", \"test\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from gtp.dataloading.tools import load_json, collect_chromosome_position_metadata\n",
    "\n",
    "\n",
    "def load_data(exp_dir, species, wing, color, chromosome):\n",
    "    position_metadata = collect_chromosome_position_metadata(\n",
    "        DNA_ROOT_DIR, species, chromosome\n",
    "    )\n",
    "    data_dir = exp_dir / f\"{species}_{wing}_{color}_chromosome_{chromosome}\"\n",
    "    all_data = []\n",
    "    for phase in [\"training\", \"validation\", \"test\"]:\n",
    "        data = np.load(\n",
    "            data_dir / f\"{phase}_windowed_editing_attributions.npy\", allow_pickle=True\n",
    "        ).item()\n",
    "        metrics = load_json(data_dir / f\"{phase}_metrics.json\")\n",
    "        pearsonr = metrics[\"pearsonr\"]\n",
    "        rmse = metrics[\"rmse\"]\n",
    "        all_data.append([data, rmse, pearsonr])\n",
    "\n",
    "    return all_data, position_metadata\n",
    "\n",
    "\n",
    "def collect_plot_data(data, nominal_positions):\n",
    "    plot_data = {}\n",
    "    for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "        all_pos = []\n",
    "        for pos in nominal_positions:\n",
    "            all_data = np.concatenate(data[pos][i])\n",
    "            all_pos.append(all_data)\n",
    "        plot_data[edit_opt] = np.stack(all_pos, axis=1)\n",
    "    return plot_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_pos_min = anno_df.Start.min()\n",
    "bp_pos_max = anno_df.End.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene', 'phenotype', 'atacSeq peak', 'exon'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno_df.Type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-20\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "\n",
    "species = \"erato\"\n",
    "wing = \"forewings\"\n",
    "color = \"color_3\"\n",
    "chromosome = 10\n",
    "all_phase_data, position_metadata = load_data(EXP_DIR, species=species, wing=wing, color=color, chromosome=10)\n",
    "\n",
    "all_nominal_positions = [list(all_phase_data[i][0].keys()) for i in range(3)]\n",
    "#assert (\n",
    "#    all_nominal_positions[0] == all_nominal_positions[1] == all_nominal_positions[2]\n",
    "#), \"All position must match across phases.\"\n",
    "\n",
    "chromosome_position_metadata = []\n",
    "chromosome_position_metadata_start_end = []\n",
    "for nom_pos in all_nominal_positions[2]:\n",
    "    if nom_pos == len(position_metadata):\n",
    "        nom_pos -= 1\n",
    "    chromosome_position_metadata.append(position_metadata[nom_pos])\n",
    "    chromosome_position_metadata_start_end.append([position_metadata[max(0, nom_pos-10)], position_metadata[min(len(position_metadata)-1, nom_pos+10)]])\n",
    "\n",
    "phase_data = all_phase_data[PHASE_NAMES.index(\"test\")]\n",
    "attrs, rmse, pearsonr = phase_data\n",
    "nominal_window_positions = list(attrs.keys())\n",
    "plot_data = collect_plot_data(attrs, nominal_window_positions)\n",
    "print(int(nominal_window_positions[0])-int(nominal_window_positions[1]))\n",
    "\n",
    "all_real_positions = [x[1] for x in chromosome_position_metadata]\n",
    "all_real_positions_start_and_end = [[x1[1], x2[1]] for x1, x2 in chromosome_position_metadata_start_end]\n",
    "\n",
    "start_pos = None\n",
    "end_pos = all_real_positions[-1]\n",
    "for i, rp in enumerate(all_real_positions):\n",
    "    if start_pos is None and rp > bp_pos_min:\n",
    "        start_pos = max(0, i-1)\n",
    "    if rp > bp_pos_max:\n",
    "        end_pos = i\n",
    "        break\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_data[\"AA\"].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = 0.0005\n",
    "max_height = 0\n",
    "fig, ax = plt.subplots(1, 1, figsize=(24, 9))\n",
    "for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "    means = np.abs(plot_data[edit_opt]).mean(axis=0)\n",
    "    y = means[start_pos:end_pos+1]\n",
    "    max_height = max(max(y), max_height)\n",
    "    for j, (s, e) in enumerate(all_real_positions_start_and_end[start_pos:end_pos+1]):\n",
    "        v = y[j]\n",
    "        rect = patches.Rectangle((s, v-(w/2)), e-s, w, linewidth=1, edgecolor='black', facecolor='purple', alpha=0.1)\n",
    "        ax.add_patch(rect)\n",
    "    ax.scatter(all_real_positions[start_pos:end_pos+1], means[start_pos:end_pos+1], label=edit_opt, s=10)\n",
    "    \n",
    "for i, row in anno_df.iterrows():\n",
    "    y = max_height + w * 10\n",
    "    if row[\"Type\"] == \"gene\":\n",
    "        y -= w*2\n",
    "        c = \"blue\"\n",
    "    elif row[\"Type\"] == \"phenotype\":\n",
    "        y -= w*4\n",
    "        c = \"yellow\"\n",
    "    elif row[\"Type\"] == \"atacSeq peak\":\n",
    "        y -= w*6\n",
    "        c = \"red\"\n",
    "    elif row[\"Type\"] == \"exon\":\n",
    "        c = \"green\"\n",
    "        y -= w*8\n",
    "    rect = patches.Rectangle((row.Start, y), row.End-row.Start, w, linewidth=1, edgecolor='black', facecolor=c)\n",
    "    ax.add_patch(rect)\n",
    "\n",
    "ax.legend()\n",
    "    \n",
    "plt.grid()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "x = gwas_df[\"ps\"]\n",
    "y = gwas_df[\"neg_log10_p\"]\n",
    "y -= y.min()\n",
    "y /= y.max()\n",
    "y *= max_height\n",
    "\n",
    "gobj = go.Scatter(\n",
    "    x=x, \n",
    "    y=y, \n",
    "    name=\"GWAS\", \n",
    "    mode='markers', \n",
    "    marker=dict(size=5, color='black'),\n",
    ")\n",
    "fig.add_trace(gobj)\n",
    "\n",
    "\n",
    "for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "    means = np.abs(plot_data[edit_opt]).mean(axis=0)\n",
    "    y = means[start_pos:end_pos+1]\n",
    "    x = all_real_positions[start_pos:end_pos+1]\n",
    "    s, e = list(zip(*all_real_positions_start_and_end[start_pos:end_pos+1]))\n",
    "    error_x = np.array(e) - np.array(x)\n",
    "    error_x_minus = np.array(x) - np.array(s)\n",
    "    gobj = go.Scatter(\n",
    "        x=x, \n",
    "        y=y, \n",
    "        name=edit_opt, \n",
    "        mode='markers', \n",
    "        marker=dict(size=5), \n",
    "        error_x=dict(type=\"data\", array=error_x, arrayminus=error_x_minus),\n",
    "    )\n",
    "    fig.add_trace(gobj)\n",
    "\n",
    "for i, row in anno_df.iterrows():\n",
    "    y = max_height + w * 10\n",
    "    if row[\"Type\"] == \"gene\":\n",
    "        y -= w*2\n",
    "        c = \"blue\"\n",
    "    elif row[\"Type\"] == \"phenotype\":\n",
    "        y -= w*4\n",
    "        c = \"yellow\"\n",
    "    elif row[\"Type\"] == \"atacSeq peak\":\n",
    "        y -= w*6\n",
    "        c = \"red\"\n",
    "    elif row[\"Type\"] == \"exon\":\n",
    "        c = \"green\"\n",
    "        y -= w*8\n",
    "    fig.add_shape(\n",
    "        type=\"rect\",\n",
    "        xref=\"x\",  # Reference x-axis data coordinates\n",
    "        yref=\"y\",  # Reference y-axis data coordinates\n",
    "        x0=row.Start,    # Starting x-coordinate of the rectangle\n",
    "        y0=y-(w/2),    # Starting y-coordinate of the rectangle\n",
    "        x1=row.End,    # Ending x-coordinate of the rectangle\n",
    "        y1=y+(w/2),    # Ending y-coordinate of the rectangle\n",
    "        line=dict(\n",
    "            color=c,\n",
    "            width=0.002,\n",
    "        ),\n",
    "        fillcolor=c,\n",
    "        opacity=1,\n",
    "        layer=\"below\" # Place the rectangle below the data points\n",
    "    )\n",
    "    \n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=[row.Start + (row.End-row.Start)/2], \n",
    "            y=[y],\n",
    "            name=row.Type,\n",
    "            hovertemplate=f\"{row.Type} {row.name} : {row.Start}-{row.End}\",\n",
    "            showlegend=False,\n",
    "            mode='markers', \n",
    "            marker=dict(color=\"black\", size=1),\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig.update_layout(dict(hovermode='closest', xaxis=dict(showspikes=True), xaxis_spikemode='across'))\n",
    "\n",
    "fig.write_html(f\"../output/results/{species}_{wing}_{color}_{chromosome}.html\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_dfs = []\n",
    "for i, edit_opt in enumerate(EDIT_OPTIONS):\n",
    "    means = np.abs(plot_data[edit_opt]).mean(axis=0)\n",
    "    y = means[start_pos:end_pos+1]\n",
    "    x = all_real_positions[start_pos:end_pos+1]\n",
    "    \n",
    "    df = pd.DataFrame(np.stack((x, y), axis=1), columns=[\"Position\", \"Attribution\"])\n",
    "    df[\"Edit_Type\"] = edit_opt\n",
    "    df.Position = df.Position.astype(int)\n",
    "    all_dfs.append(df)\n",
    "\n",
    "all_df = pd.concat(all_dfs)\n",
    "\n",
    "all_df.to_csv(f\"../output/results/{species}_{wing}_{color}_{chromosome}.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
