{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocess Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import socket\n",
    "from enum import Enum\n",
    "\n",
    "\n",
    "class Hosts(Enum):\n",
    "    IMAGEOMICS_SERVER = \"cse-cnc196909s.coeit.osu.edu\"\n",
    "\n",
    "\n",
    "hostname = socket.gethostname()\n",
    "if hostname == Hosts.IMAGEOMICS_SERVER.value:\n",
    "    ROOT_PHENOTYPE_INPUT_DIR = Path(\"/local/scratch/carlyn.1/dna/colors\")\n",
    "    ROOT_PHENOTYPE_OUTPUT_DIR = Path(\"/local/scratch/carlyn.1/dna/processed/phenotypes\")\n",
    "    ROOT_GENOTYPE_INPUT_DIR = Path(\"/local/scratch/carlyn.1/dna/vcfs\")\n",
    "    ROOT_GENOTYPE_OUTPUT_DIR = Path(\"/local/scratch/carlyn.1/dna/processed\")\n",
    "else:\n",
    "    ROOT_PHENOTYPE_INPUT_DIR = Path(\"/local/scratch/david/geno-pheno-data/colors\")\n",
    "    ROOT_PHENOTYPE_OUTPUT_DIR = Path(\n",
    "        \"/local/scratch/david/geno-pheno-data/colors/processed\"\n",
    "    )\n",
    "    ROOT_GENOTYPE_INPUT_DIR = Path(\"/local/scratch/david/geno-pheno-data/dna/\")\n",
    "    ROOT_GENOTYPE_OUTPUT_DIR = Path(\n",
    "        \"/local/scratch/david/geno-pheno-data/dna/processed\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gtp.dataloading.data_preprocessors import ButterflyPatternizePreprocessor\n",
    "\n",
    "preprocessor = ButterflyPatternizePreprocessor(\n",
    "    input_dir=ROOT_PHENOTYPE_INPUT_DIR, output_dir=ROOT_PHENOTYPE_OUTPUT_DIR\n",
    ")\n",
    "\n",
    "for species in [\"erato\", \"melpomene\"]:\n",
    "    for wing in [\"forewings\", \"hindwings\"]:\n",
    "        for color in [\"color_1\", \"color_2\", \"color_3\", \"total\", \"bioclip\"]:\n",
    "            suffix_path = f\"{species}_{wing}_PCA/PCA_{color}_loadings.csv\"\n",
    "            preprocessor.process(pca_csv_path_suffix=suffix_path)\n",
    "            preprocessor.save_result(f\"{species}_{wing}_{color}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from enum import Enum\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "from tqdm import tqdm\n",
    "from pathlib import Path\n",
    "\n",
    "from gtp.dataloading.data_preprocessors import ButterflyGenePreprocessor\n",
    "from gtp.tools.simple import convert_bytes, get_filesize\n",
    "\n",
    "\n",
    "class DNA_SCOPE(Enum):\n",
    "    GENOME = \"genome\"\n",
    "    CHROMOSOME = \"chromosomes\"\n",
    "    GENE = \"genes\"\n",
    "\n",
    "\n",
    "THIS_DNA_SCOPE = DNA_SCOPE.GENOME.value\n",
    "\n",
    "\n",
    "def process_fn(pca_csv_path_suffix, save_dir):\n",
    "    size = convert_bytes(get_filesize(ROOT_GENOTYPE_INPUT_DIR / pca_csv_path_suffix))\n",
    "    #print(f\"Processing {pca_csv_path_suffix}: {size} bytes\")\n",
    "    preprocessor = ButterflyGenePreprocessor(\n",
    "        input_dir=ROOT_GENOTYPE_INPUT_DIR, output_dir=ROOT_GENOTYPE_OUTPUT_DIR, save_format=\"parquet\"\n",
    "    )\n",
    "    preprocessor.process(pca_csv_path_suffix=pca_csv_path_suffix, processor=\"pandas\")\n",
    "    preprocessor.save_result(save_dir)\n",
    "    #print(f\"Completed {pca_csv_path_suffix}: {size} bytes\")\n",
    "    return True\n",
    "\n",
    "\n",
    "futures = []\n",
    "process_data = []\n",
    "pool = ThreadPoolExecutor(1)\n",
    "\n",
    "for species in [\"erato\", \"melpomene\"]:\n",
    "    species_genome_path = Path(f\"{species}/{THIS_DNA_SCOPE}\")\n",
    "    #print(ROOT_GENOTYPE_INPUT_DIR / species_genome_path)\n",
    "    for root, dirs, files in os.walk(ROOT_GENOTYPE_INPUT_DIR / species_genome_path):\n",
    "        for i, f in enumerate(files):\n",
    "            fname = f.split(\".\")[0]\n",
    "            #if os.path.exists(\n",
    "            #    ROOT_GENOTYPE_OUTPUT_DIR\n",
    "            #    / f\"{THIS_DNA_SCOPE}/{species}/{fname}/ml_ready.npy\"\n",
    "            #):\n",
    "            #    continue  # Delete this after\n",
    "            \n",
    "            genome_file_path = species_genome_path / f\n",
    "            size = os.path.getsize(ROOT_GENOTYPE_INPUT_DIR / genome_file_path)\n",
    "            process_data.append(\n",
    "                [size, species, genome_file_path, f\"{THIS_DNA_SCOPE}/{species}/{fname}\"]\n",
    "            )\n",
    "\n",
    "process_data = sorted(process_data, key=lambda x: (x[1], x[0]))\n",
    "\n",
    "\n",
    "pbar = tqdm(process_data, desc=\"Processing Genotype data.\")\n",
    "for size, species, genome_file_path, save_dir in pbar:\n",
    "    size = convert_bytes(os.path.getsize(ROOT_GENOTYPE_INPUT_DIR / genome_file_path))\n",
    "    pbar.set_description_str(f\"Processing file of size: {size}\")\n",
    "    process_fn(genome_file_path, save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from enum import Enum\n",
    "from tqdm import tqdm\n",
    "\n",
    "class DNA_SCOPE(Enum):\n",
    "    GENOME = \"genome\"\n",
    "    CHROMOSOME = \"chromosomes\"\n",
    "    GENE = \"genes\"\n",
    "\n",
    "\n",
    "THIS_DNA_SCOPE = DNA_SCOPE.GENOME.value\n",
    "\n",
    "species = \"melpomene\"\n",
    "data_dir = Path(ROOT_GENOTYPE_OUTPUT_DIR, \"genotypes\", \"genome\", species)\n",
    "path_data = []\n",
    "for root, dirs, files in data_dir.walk():\n",
    "    if \"ml_ready.npy\" in files:\n",
    "        scaffold = root.parts[-1]\n",
    "        if species == \"erato\":\n",
    "            chrom_info = scaffold.replace(\"Herato\", \"\")\n",
    "            scaf_num = int(chrom_info)\n",
    "        elif species == \"melpomene\":\n",
    "            chrom_info = scaffold.replace(\"Hmel2\", \"\")\n",
    "            scaf_num = int(chrom_info[:-1])\n",
    "        \n",
    "        path_data.append([root, scaf_num])\n",
    "    \n",
    "path_data = sorted(path_data, key=lambda x: x[1])\n",
    "    \n",
    "prev_camids = None\n",
    "\n",
    "output_dir = Path(\"/local/scratch/carlyn.1/dna/processed/genotypes/genome_individuals\", species)\n",
    "#output_dir = Path(f\"/local/scratch/carlyn.1/dna/vcfs/{species}/genome_individuals\")\n",
    "output_dir.mkdir(exist_ok=True)\n",
    "for root, _ in tqdm(path_data, desc=\"Splitting data per individual\", colour=\"#9E2776\"):\n",
    "    if \"ml_ready.npy\" in files:\n",
    "        # Assert aligned CAMIDS\n",
    "        camids = np.load(root / \"camids.npy\")\n",
    "        if prev_camids is not None:\n",
    "            assert (camids == prev_camids).all()\n",
    "        prev_camids = camids\n",
    "        \n",
    "        ml_data = np.load(root / \"ml_ready.npy\")\n",
    "        \n",
    "        for camid, row in zip(camids, ml_data):\n",
    "            tgt_file = output_dir / f\"{camid}.npy\"\n",
    "            if tgt_file.exists():\n",
    "                ind_data = np.load(tgt_file)\n",
    "                ind_data = np.concatenate((ind_data, row), axis=0)\n",
    "                np.save(tgt_file, ind_data)\n",
    "            else:\n",
    "                np.save(tgt_file, row)\n",
    "        \n",
    "            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gtp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
